// @canva/app-ui-kit/dist/cjs/ui/base/swatch/internal/swatch_util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useContrastingColors = exports.gradientToCssBackground = exports.generateSwatchLabel = exports.getValidGradientString = exports.toStripeWidth = exports.toStripePosition = exports.isFillObject = void 0;
const preconditions_1 = require('../../../../base/preconditions');
const unique_1 = require('../../../../base/unique');
const color_1 = require('../../color/color');
const escape_hatch_1 = require('../../theme/escape_hatch');
const swatch_messages_1 = require("../swatch.messages");
function isFillObject(color) {
    if (typeof color === 'object' && color != null && 'type' in color) {
        return color.type === 'linear' || color.type === 'radial';
    }
    return false;
}
exports.isFillObject = isFillObject;
const toStripePosition = (index, totalStripes) => {
    const stripePercentage = totalStripes > 0 ? 100 / totalStripes : 0;
    const positionPercentage = index * stripePercentage;
    return `${positionPercentage.toFixed(2)}%`;
};
exports.toStripePosition = toStripePosition;
const toStripeWidth = (width, listSize) => {
    const swatchWidth = width > 0 ? width : 0;
    const stripeWidth = listSize > 0 ? swatchWidth / listSize : swatchWidth;
    return `${Math.ceil(stripeWidth)}px`;
};
exports.toStripeWidth = toStripeWidth;
const isGradient = (color) => /gradient/i.test(color);
const getValidGradientString = (color) => isGradient(color) ? '' : `linear-gradient(to right, ${color}, ${color})`;
exports.getValidGradientString = getValidGradientString;
const normalizeValue = (v) => {
    return v.toFixed(3).replace(/\.?0*$/, '');
};
const convertToPercentage = (n) => {
    return n <= 1 ? `${normalizeValue(n * 100)}%` : `${n}%`;
};
const convertColorToHex = (color) => {
    const { r, g, b } = color_1.CssColors.fromString(color);
    return color_1.Conversion.toHex(new color_1.RgbColor(r, g, b));
};
const generateGradientLabel = (color) => {
    const stops = color.stops
        .map(c => c.transparency
        ? swatch_messages_1.SwatchMessages.colorWithTransparencyLabel(convertColorToHex(c.color), convertToPercentage(c.transparency))
        : convertColorToHex(c.color))
        .join(`, `);
    switch (color.type) {
        case 'linear': {
            return swatch_messages_1.SwatchMessages.colorLinearGradientLabel(color.rotation, stops);
        }
        case 'radial': {
            const { center: { top: y, left: x }, } = color;
            return swatch_messages_1.SwatchMessages.colorCircularRadialGradientLabel(`${convertToPercentage(x)} ${convertToPercentage(y)}`, stops);
        }
        default:
            throw new preconditions_1.UnreachableError(color);
    }
};
const generateSwatchLabel = ({ areAllColorsTransparent, colors, colorNames, rgbToCmyk, fill, isAtleastOneColorTransparent, areNoColorsApplied, }) => {
    if (areAllColorsTransparent || areNoColorsApplied) {
        return swatch_messages_1.SwatchMessages.colorTransparentLabel();
    }
    if (Array.isArray(colors) && colors.length === 1 && typeof colors[0] === 'string') {
        return getColorTooltipLabel(colors[0], colorNames, rgbToCmyk);
    }
    if (Array.isArray(fill)) {
        const generatedLabel = fill
            .map(color => {
            if (typeof color === 'string') {
                return getColorTooltipLabel(color);
            }
            if (color == null) {
                return undefined;
            }
            if (isFillObject(color)) {
                return generateGradientLabel(color);
            }
            return undefined;
        })
            .filter(Boolean)
            .join(', ');
        return isAtleastOneColorTransparent
            ? swatch_messages_1.SwatchMessages.colorSemiTransparentLabel(generatedLabel)
            : generatedLabel || swatch_messages_1.SwatchMessages.invalidColor();
    }
    return swatch_messages_1.SwatchMessages.invalidColor();
};
exports.generateSwatchLabel = generateSwatchLabel;
const convertColorToRgba = (color, transparency = 0) => {
    const { r, g, b } = color_1.CssColors.fromString(color);
    return new color_1.RgbColor(r, g, b).toRgbaColor(1 - transparency).toRgbaString();
};
const gradientToCssBackground = (color) => {
    const colorStops = color.stops
        .map(({ color, position, transparency = 0 }) => {
        const rgbaColor = convertColorToRgba(color, transparency);
        return position ? `${rgbaColor} ${convertToPercentage(position)}` : `${rgbaColor}`;
    })
        .join(', ');
    switch (color.type) {
        case 'linear': {
            return `linear-gradient(${color.rotation}deg, ${colorStops})`;
        }
        case 'radial': {
            const { center: { top: y, left: x }, } = color;
            const center = `${convertToPercentage(x)} ${convertToPercentage(y)}`;
            return `radial-gradient(circle farthest-corner at ${center}, ${colorStops})`;
        }
        default:
            throw new preconditions_1.UnreachableError(color);
    }
};
exports.gradientToCssBackground = gradientToCssBackground;
const RGBA_RE = /^rgba\((\d+),\s*(\d+),\s*(\d+),\s*([0-9]*\.?[0-9]+)\)$/;
const getColorTooltipLabel = (color, colorNames, rgbToCmyk) => {
    var _a;
    const cmyk = (_a = rgbToCmyk === null || rgbToCmyk === void 0 ? void 0 : rgbToCmyk.get(color)) === null || _a === void 0 ? void 0 : _a.split(',').join(', ');
    const name = colorNames === null || colorNames === void 0 ? void 0 : colorNames.get(color);
    const cmykMessage = cmyk && swatch_messages_1.SwatchMessages.cmykInfo(cmyk);
    if (cmykMessage && name) {
        return swatch_messages_1.SwatchMessages.colorTitleWithNameAndCmykInfo(name, color, cmykMessage);
    }
    else if (name) {
        return swatch_messages_1.SwatchMessages.colorTitleWithName(name, color);
    }
    else if (cmykMessage) {
        return swatch_messages_1.SwatchMessages.colorTitleWithCmykInfo(color, cmykMessage);
    }
    else if (color === 'circularSpectrum' || color === 'linearSpectrum') {
        return swatch_messages_1.SwatchMessages.colorRangeLabel();
    }
    else if (color.match(RGBA_RE)) {
        const { a } = color_1.CssColors.fromRgbaString(color);
        const transparency = 1 - a;
        if (transparency > 0) {
            return swatch_messages_1.SwatchMessages.colorWithTransparencyLabel(convertColorToHex(color), convertToPercentage(transparency));
        }
        return convertColorToHex(color);
    }
    try {
        return convertColorToHex(color);
    }
    catch {
        return color;
    }
};
function useContrastingColors(colors, shouldContrastColor) {
    const colorList = colors === 'circularSpectrum' || colors === 'linearSpectrum' ? [colors] : colors;
    const uniqueColors = (0, unique_1.unique)(colorList);
    const themeColor = (0, escape_hatch_1.useThemedMedia)({
        value: { light: '#ffffff', dark: '#000000' },
    });
    const nonContrastColor = themeColor || { value: '#ffffff' };
    let themeColors;
    if (nonContrastColor.value === '#ffffff') {
        themeColors = ['#ffffff', 'white', '#fff'];
    }
    else {
        themeColors = ['#000000', 'black', '#000'];
    }
    const uniqueColor = Array.isArray(uniqueColors) && uniqueColors.length === 1 && typeof uniqueColors[0] === 'string'
        ? uniqueColors[0]
        : null;
    if (shouldContrastColor && uniqueColor && themeColors.includes(uniqueColor)) {
        return 'circularSpectrum';
    }
    return uniqueColors;
}
exports.useContrastingColors = useContrastingColors;
