// @canva/app-ui-kit/dist/cjs/ui/base/bootstrap/ui_bootstrap_proto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UiBootstrap = exports.AutoplayVideosUtil = exports.AutoplayVideos = exports.ThemeUtil = exports.Theme = exports.DirectionUtil = exports.Direction = void 0;
const proto_1 = require('../../../base/proto/proto');
var Direction;
(function (Direction) {
    Direction[Direction["LTR"] = 1] = "LTR";
    Direction[Direction["RTL"] = 2] = "RTL";
})(Direction || (exports.Direction = Direction = {}));
exports.DirectionUtil = proto_1.Proto.createEnumUtil(() => [
    1,
    2,
], 1);
var Theme;
(function (Theme) {
    Theme[Theme["ADAPTIVE"] = 1] = "ADAPTIVE";
    Theme[Theme["LIGHT"] = 2] = "LIGHT";
    Theme[Theme["DARK"] = 3] = "DARK";
})(Theme || (exports.Theme = Theme = {}));
exports.ThemeUtil = proto_1.Proto.createEnumUtil(() => [
    1, 'ADAPTIVE',
    2, 'LIGHT',
    3, 'DARK',
]);
var AutoplayVideos;
(function (AutoplayVideos) {
    AutoplayVideos[AutoplayVideos["ADAPTIVE"] = 1] = "ADAPTIVE";
    AutoplayVideos[AutoplayVideos["DONT_AUTOPLAY"] = 2] = "DONT_AUTOPLAY";
    AutoplayVideos[AutoplayVideos["AUTOPLAY"] = 3] = "AUTOPLAY";
})(AutoplayVideos || (exports.AutoplayVideos = AutoplayVideos = {}));
exports.AutoplayVideosUtil = proto_1.Proto.createEnumUtil(() => [
    0,
    1,
    2,
]);
exports.UiBootstrap = proto_1.Proto.createMessage(() => ({
    enableAnimations: proto_1.Proto.requiredBoolean(1),
    direction: proto_1.Proto.requiredStringEnum(2, exports.DirectionUtil),
    disableFocusTraps: proto_1.Proto.requiredBoolean(3),
    enableUserSelection: proto_1.Proto.requiredBoolean(4),
    theme: proto_1.Proto.optionalStringEnum(7, exports.ThemeUtil),
    enableHighColorContrast: proto_1.Proto.optionalBoolean(8),
    enableEaselUiRefresh: proto_1.Proto.requiredBoolean(13),
    autoplayVideos: proto_1.Proto.optionalStringEnum(11, exports.AutoplayVideosUtil),
    disableDialogBlur: proto_1.Proto.requiredBoolean(9),
    exposeStringIds: proto_1.Proto.requiredBoolean(10),
}));
