// @canva/app-ui-kit/dist/cjs/ui/base/gestures/pointer_tracker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PointerTracker = void 0;
const mouse_tracker_1 = require("./mouse_tracker");
const touch_tracker_1 = require("./touch_tracker");
const DEDUPE_TIMEOUT = 2500;
const DEDUPE_DISTANCE = 25;
class PointerTracker {
    constructor(element, onPointerDown, onPointerMove, onPointerUp, onPointerCancel) {
        this.recentTouchEvents = new Set();
        this.mouseTracker = new mouse_tracker_1.MouseTracker(element, this.onMouseEvent(onPointerDown), this.onMouseEvent(onPointerMove), this.onMouseEvent(onPointerUp));
        this.touchTracker = new touch_tracker_1.TouchTracker(element, this.onTouchEvent(onPointerDown), this.onTouchEvent(onPointerMove), this.onTouchEvent(onPointerUp), this.onTouchEvent(onPointerCancel));
    }
    onTouchEvent(handler) {
        return (pointer) => {
            this.recentTouchEvents.add(pointer);
            setTimeout(() => this.recentTouchEvents.delete(pointer), DEDUPE_TIMEOUT);
            handler(pointer);
        };
    }
    onMouseEvent(handler) {
        return (pointer) => {
            if (!this.isSynthetic(pointer)) {
                handler(pointer);
            }
        };
    }
    isSynthetic(mouse) {
        var _a;
        if ((_a = mouse.sourceEvent.sourceCapabilities) === null || _a === void 0 ? void 0 : _a.firesTouchEvents) {
            return true;
        }
        for (const touch of this.recentTouchEvents) {
            const dx = Math.abs(mouse.x - touch.x);
            const dy = Math.abs(mouse.y - touch.y);
            if (dx <= DEDUPE_DISTANCE && dy <= DEDUPE_DISTANCE) {
                return true;
            }
        }
    }
    disconnect() {
        this.mouseTracker.disconnect();
        this.touchTracker.disconnect();
    }
}
exports.PointerTracker = PointerTracker;
