// @canva/app-ui-kit/dist/esm/ui/base/bootstrap/ui_bootstrap_proto.js
import { Proto } from '../../../base/proto/proto';
export const Direction = {
  LTR: 1,
  RTL: 2
};
export const DirectionUtil = Proto.createEnumUtil(() => [1, 2], 1);
export const Theme = {
  ADAPTIVE: 1,
  LIGHT: 2,
  DARK: 3
};
export const ThemeUtil = Proto.createEnumUtil(() => [1, 'ADAPTIVE', 2, 'LIGHT', 3, 'DARK']);
export const AutoplayVideos = {
  ADAPTIVE: 1,
  DONT_AUTOPLAY: 2,
  AUTOPLAY: 3
};
export const AutoplayVideosUtil = Proto.createEnumUtil(() => [0, 1, 2]);
export const UiBootstrap = Proto.createMessage(() => ({
  enableAnimations: Proto.requiredBoolean( 1),
  direction: Proto.requiredStringEnum( 2, DirectionUtil),
  disableFocusTraps: Proto.requiredBoolean( 3),
  enableUserSelection: Proto.requiredBoolean( 4),
  theme: Proto.optionalStringEnum( 7, ThemeUtil),
  enableHighColorContrast: Proto.optionalBoolean( 8),
  enableEaselUiRefresh: Proto.requiredBoolean( 13),
  autoplayVideos: Proto.optionalStringEnum( 11, AutoplayVideosUtil),
  disableDialogBlur: Proto.requiredBoolean( 9),
  exposeStringIds: Proto.requiredBoolean( 10)
}));
