// @canva/app-ui-kit/dist/cjs/ui/base/provider/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useEnableEaselUiRefresh = exports.useDisableDialogBlur = exports.useAutoplayVideos = exports.useEnableHighColorContrast = exports.useTheme = exports.useEnableUserSelection = exports.useDisableFocusTraps = exports.useDirection = exports.useEnableAnimations = exports.EaselProvider = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const preconditions_1 = require('../../../base/preconditions');
const React = require("react");
const EaselContext = React.createContext(null);
function EaselProvider({ enableAnimations = true, direction, disableFocusTraps, enableUserSelection, theme, enableHighColorContrast, autoplayVideos, disableDialogBlur, enableEaselUiRefresh, children, }) {
    return ((0, jsx_runtime_1.jsx)(EaselContext.Provider, { value: {
            enableAnimations,
            direction,
            disableFocusTraps,
            enableUserSelection,
            theme,
            enableHighColorContrast,
            autoplayVideos,
            disableDialogBlur,
            enableEaselUiRefresh,
        }, children: children }));
}
exports.EaselProvider = EaselProvider;
function useEaselContext() {
    return preconditions_1.Preconditions.checkExists(React.useContext(EaselContext), 'Did you forget to render a provider?');
}
function useEnableAnimations() {
    return useEaselContext().enableAnimations;
}
exports.useEnableAnimations = useEnableAnimations;
function useDirection() {
    return useEaselContext().direction;
}
exports.useDirection = useDirection;
function useDisableFocusTraps() {
    return useEaselContext().disableFocusTraps;
}
exports.useDisableFocusTraps = useDisableFocusTraps;
function useEnableUserSelection() {
    return useEaselContext().enableUserSelection;
}
exports.useEnableUserSelection = useEnableUserSelection;
function useTheme() {
    return useEaselContext().theme;
}
exports.useTheme = useTheme;
function useEnableHighColorContrast() {
    return useEaselContext().enableHighColorContrast;
}
exports.useEnableHighColorContrast = useEnableHighColorContrast;
function useAutoplayVideos() {
    return useEaselContext().autoplayVideos;
}
exports.useAutoplayVideos = useAutoplayVideos;
function useDisableDialogBlur() {
    return useEaselContext().disableDialogBlur;
}
exports.useDisableDialogBlur = useDisableDialogBlur;
function useEnableEaselUiRefresh() {
    return useEaselContext().enableEaselUiRefresh;
}
exports.useEnableEaselUiRefresh = useEnableEaselUiRefresh;
