// @canva/app-ui-kit/dist/esm/ui/base/button/internal/repeat_click_wrapper.js
import { makeObservable } from '../../../../base/make_observable/make_observable';
import { __decorate } from "tslib";
import { action, observable, runInAction } from 'mobx';
import { observer } from 'mobx-react';
import * as React from 'react';
import { isPointerEventsSupported } from '../../device_capabilities/device_capabilities';
export const SLOW_REPEAT_DELAY = 450;
export const FAST_REPEAT_DELAY = 350;
export const SLOW_REPEAT_INTERVAL = 100;
export const FAST_REPEAT_INTERVAL = 40;
let RepeatClickWrapper = class RepeatClickWrapper extends React.Component {
  static _makeObservable(instance) {
    makeObservable(instance, {
      repeatDelayTimeoutId: observable.ref,
      repeatIntervalId: observable.ref,
      clearTimers: action.bound,
      onPointerDown: action.bound
    });
  }
  constructor() {
    super(...arguments);
    RepeatClickWrapper._makeObservable(this);
    this.isPointerDown = false;
    this.onPointerUp = e => {
      if (this.isPointerDown) {
        this.isPointerDown = false;
        if (!this.repeatIntervalId) {
          this.maybeClick(e);
        }
        this.clearTimers();
      }
      e.preventDefault();
    };
    this.onPointerLeave = e => {
      this.isPointerDown = false;
      this.clearTimers();
      e.preventDefault();
    };
    this.preventContextMenu = e => {
      e.preventDefault();
      return false;
    };
  }
  componentWillUnmount() {
    this.clearTimers();
  }
  clearTimers() {
    if (this.repeatDelayTimeoutId) {
      window.clearTimeout(this.repeatDelayTimeoutId);
      this.repeatDelayTimeoutId = undefined;
    }
    if (this.repeatIntervalId) {
      window.clearInterval(this.repeatIntervalId);
      this.repeatIntervalId = undefined;
    }
  }
  onPointerDown(e) {
    this.isPointerDown = true;
    const {
      repeatSpeed
    } = this.props;
    const repeatInterval = repeatSpeed === 'fast' ? FAST_REPEAT_INTERVAL : SLOW_REPEAT_INTERVAL;
    const repeatDelay = repeatSpeed === 'fast' ? FAST_REPEAT_DELAY : SLOW_REPEAT_DELAY;
    this.clearTimers();
    this.repeatDelayTimeoutId = window.setTimeout(() => {
      runInAction(() => {
        this.maybeClick(e);
        this.repeatIntervalId = window.setInterval(() => {
          this.maybeClick(e);
        }, repeatInterval);
      });
    }, repeatDelay);
  }
  maybeClick(e) {
    var _a, _b;
    if (!this.props.disabled) {
      (_b = (_a = this.props).onClick) === null || _b === void 0 ? void 0 : _b.call(_a, e);
    }
  }
  render() {
    if (this.props.repeatSpeed !== 'none' && isPointerEventsSupported()) {
      return this.props.children({
        onPointerDown: this.onPointerDown,
        onPointerUp: this.onPointerUp,
        onPointerLeave: this.onPointerLeave,
        onContextMenu: this.preventContextMenu,
        onClick: undefined
      });
    }
    return this.props.children({
      onClick: this.props.onClick
    });
  }
};
RepeatClickWrapper.defaultProps = {
  repeatSpeed: 'none'
};
RepeatClickWrapper = __decorate([observer], RepeatClickWrapper);
export { RepeatClickWrapper };
