// @canva/app-ui-kit/dist/esm/ui/base/color_picker/color_picker.js
import { makeObservable } from '../../../base/make_observable/make_observable';
import { __decorate } from "tslib";
import { jsx as _jsx } from "react/jsx-runtime";
import * as mobx from 'mobx';
import * as mobxReact from 'mobx-react';
import * as React from 'react';
import { ColorPickerPresenter, ColorPickerStore, ColorStoreImpl } from './color_picker_presenter';
import { InternalColorPicker } from './internal/internal_color_picker';
let ColorPicker = class ColorPicker extends React.Component {
  static _makeObservable(instance) {
    makeObservable(instance, {
      componentDidUpdate: mobx.action
    });
  }
  constructor(props) {
    super(props);
    ColorPicker._makeObservable(this);
    const alpha = props.alpha != null ? props.alpha : 1;
    this.colorStore = new ColorStoreImpl(props.color, alpha);
    this.store = new ColorPickerStore(this.colorStore);
    this.presenter = new ColorPickerPresenter(this.store, props.onChange);
  }
  componentDidUpdate(previousProps) {
    if (this.props.onChange !== previousProps.onChange) {
      this.presenter = new ColorPickerPresenter(this.store, this.props.onChange);
    }
    if (this.props.color !== previousProps.color || this.props.alpha !== previousProps.alpha) {
      this.colorStore.color = this.props.color;
      this.colorStore.alpha = this.props.alpha != null ? this.props.alpha : 1;
    }
  }
  render() {
    const {
      sizingMode,
      isSheet,
      autoFocusHexInput,
      alpha,
      onChangeStart,
      onChangeComplete,
      onStartEyedropper,
      onDeleteColor
    } = this.props;
    const {
      onInputChange,
      onHueChange,
      onSatValChange,
      onAlphaChange
    } = this.presenter;
    const {
      color
    } = this.store;
    return _jsx(InternalColorPicker, {
      color: color,
      sizingMode: sizingMode,
      isSheet: isSheet,
      autoFocusHexInput: autoFocusHexInput,
      onChangeStart: onChangeStart,
      onChangeComplete: onChangeComplete,
      onInputChange: onInputChange,
      onHueChange: onHueChange,
      onSatValChange: onSatValChange,
      onAlphaChange: alpha != null ? onAlphaChange : undefined,
      onStartEyedropper: onStartEyedropper,
      onDeleteColor: onDeleteColor
    });
  }
};
ColorPicker = __decorate([mobxReact.observer], ColorPicker);
export { ColorPicker };
