// @canva/app-ui-kit/dist/esm/ui/base/layout/rows/rows.js
import classNames from 'classnames';
import * as React from 'react';
import { commonPropsToDOMProps } from '../../box/types';
import { getSpaceValue } from '../../metrics/metrics';
import { generateStyle } from '../../responsive/responsive';
import styles, { customProperties } from './rows.css';
const spacingPropertyMap = {
    default: customProperties.spacing,
    smallUp: customProperties.smallUpSpacing,
    mediumUp: customProperties.mediumUpSpacing,
    largeUp: customProperties.largeUpSpacing,
};
const alignPropertyMap = {
    default: customProperties.align,
    smallUp: customProperties.smallUpAlign,
    mediumUp: customProperties.mediumUpAlign,
    largeUp: customProperties.largeUpAlign,
};
export const Rows = React.forwardRef(function Rows({ children, spacing, align, tagName = 'div', ...rest }, ref) {
    return React.createElement(tagName, {
        className: classNames(styles.rows),
        style: {
            ...(spacing && generateStyle(spacingPropertyMap, spacing, getSpaceValue)),
            ...(align && generateStyle(alignPropertyMap, align)),
        },
        children,
        ref,
        ...commonPropsToDOMProps(rest),
    });
});
Rows.displayName = 'Rows';
