// @canva/app-ui-kit/dist/esm/ui/base/media_query/media_query_presenter.js
import { makeObservable } from '../../../base/make_observable/make_observable';
import { __decorate } from "tslib";
import { debounce } from '../../../base/debounce';
import { action, observable } from 'mobx';
const MEDIA_QUERY_LISTENER_DEBOUNCE_TIME = 16;
export class MediaQueryStore {
  static _makeObservable(instance) {
    makeObservable(instance, {
      matchedQueries: observable.ref
    });
  }
  constructor() {
    MediaQueryStore._makeObservable(this);
    this.matchedQueries = [];
  }
}
export class MediaQueryPresenter {
  static _makeObservable(instance) {
    makeObservable(instance, {
      updateQueries: action
    });
  }
  constructor({
    matchMedia
  }) {
    MediaQueryPresenter._makeObservable(this);
    this.mqls = [];
    this.matchMedia = matchMedia || window && window.matchMedia;
  }
  getMatches(queries) {
    if (!this.matchMedia) {
      return [];
    }
    const matchMedia = this.matchMedia;
    return queries.filter(query => matchMedia(query).matches);
  }
  updateQueries(queries, store) {
    this.registerListener(queries, store);
    store.matchedQueries = this.getMatches(queries);
  }
  registerListener(queries, store) {
    this.clear();
    if (!this.matchMedia) {
      return;
    }
    const matchMedia = this.matchMedia;
    const listener = debounce(action(() => store.matchedQueries = this.getMatches(queries)), MEDIA_QUERY_LISTENER_DEBOUNCE_TIME);
    this.listener = listener;
    this.mqls = queries.map(query => {
      const mql = matchMedia(query);
      if (typeof mql.addEventListener === 'function') {
        mql.addEventListener('change', listener);
      } else {
        mql.addListener(listener);
      }
      return mql;
    });
  }
  clear() {
    if (this.listener) {
      const listener = this.listener;
      this.mqls.forEach(mql => {
        if (typeof mql.removeEventListener === 'function') {
          mql.removeEventListener('change', listener);
        } else {
          mql.removeListener(listener);
        }
      });
    }
  }
}
