// @canva/app-ui-kit/dist/cjs/ui/base/sheet/internal_sheet.js
"use strict";

const makeObservable = require('../../../base/make_observable/make_observable').makeObservable;
var InternalSheet_1;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InternalSheet = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const id_generator_1 = require('../../../base/id_generator/id_generator');
const classnames_1 = require("classnames");
const mobx_1 = require("mobx");
const mobx_react_1 = require("mobx-react");
const React = require("react");
const device_capabilities_1 = require('../device_capabilities/device_capabilities');
const layer_1 = require('../layer/layer');
const mobile_event_handler_1 = require('../mobile_event_handler/mobile_event_handler');
const theme_1 = require('../theme/theme');
const layer_view_1 = require("./internal/layer_view");
const sheet_css_1 = require("./internal/sheet.css");
const defaultPlaceholder = () => null;
let InternalSheet = InternalSheet_1 = class InternalSheet extends React.Component {
  static _makeObservable(instance) {
    makeObservable(instance, {
      contentElement: mobx_1.observable.ref,
      outerElement: mobx_1.observable.ref,
      parentLayerContainer: mobx_1.observable.ref,
      show: mobx_1.observable.ref,
      currentContent: mobx_1.observable.ref,
      currentOuterContent: mobx_1.observable.ref,
      themeData: mobx_1.observable.ref,
      themeClass: mobx_1.computed,
      window: mobx_1.observable.ref,
      supportsSafeAreaInset: mobx_1.computed,
      setThemeData: mobx_1.action.bound,
      componentDidMount: mobx_1.action,
      componentDidUpdate: mobx_1.action
    });
  }
  get themeClass() {
    var _a;
    return (_a = this.themeData) === null || _a === void 0 ? void 0 : _a.className;
  }
  get supportsSafeAreaInset() {
    return this.window ? (0, device_capabilities_1.supportsSafeAreaInsetValues)(this.window.navigator) : undefined;
  }
  constructor(props) {
    super(props);
    InternalSheet._makeObservable(this);
    this.id = InternalSheet_1.ids.next();
    this.closeTimeout = 0;
    this.onBackButton = event => {
      const {
        onBackButton,
        open
      } = this.props;
      onBackButton && open && onBackButton(event);
    };
    if (props.sheetBehavior) {
      this.sheetDragInternal = {
        sheetBehavior: props.sheetBehavior
      };
    }
    this.easelConfiguration = props.easelConfiguration;
    this.show = !!this.props.open;
    if (this.show || props.mountBehaviour === 'eager') {
      this.currentContent = props.content;
      this.currentOuterContent = props.outerContent;
    }
  }
  render() {
    return (0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, {
      children: [(0, jsx_runtime_1.jsx)(theme_1.WithThemeData, {
        children: this.setThemeData
      }), (0, jsx_runtime_1.jsx)(mobile_event_handler_1.MobileEventHandler, {
        onBackButton: this.onBackButton,
        children: (0, jsx_runtime_1.jsx)(layer_1.WithLayerParent, {
          parentLayer: this.parentLayerContainer,
          children: this.contentElement && this.outerElement && (0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, {
            children: [this.renderPortal(this.currentContent, this.contentElement), this.renderPortal(this.currentOuterContent, this.outerElement)]
          })
        })
      })]
    });
  }
  setThemeData(data) {
    this.themeData = data;
    return null;
  }
  renderPortal(content, element) {
    if (!content) {
      return null;
    }
    const Content = typeof content === 'function' ? content() : content;
    return this.props.createPortal(Content, element);
  }
  componentDidMount() {
    var _a;
    this.window = (_a = this.props.windowOverride) !== null && _a !== void 0 ? _a : window;
    this.outerElement = this.window.document.createElement('div');
    this.contentElement = this.window.document.createElement('div');
    this.outerElement.className = (0, classnames_1.default)(sheet_css_1.default.sheetElement, sheet_css_1.default.scrollable);
    this.contentElement.className = (0, classnames_1.default)(sheet_css_1.default.sheetElement, {
      [sheet_css_1.default.scrollable]: this.props.scrollable
    });
    this.contentElement.style.display = this.props.open ? '' : 'none';
    if (this.currentContent) {
      this.props.layers.addSheet(this);
    }
    window.clearTimeout(this.closeTimeout);
    (0, mobx_react_1.disposeOnUnmount)(this, [
    (0, mobx_1.reaction)(() => {
      var _a, _b;
      return (_b = (_a = this.themeData) === null || _a === void 0 ? void 0 : _a.classNames) !== null && _b !== void 0 ? _b : [];
    }, (next, previous) => {
      if (this.outerElement == null || this.contentElement == null) {
        return;
      }
      if (previous != null) {
        this.contentElement.classList.remove(...previous);
        this.outerElement.classList.remove(...previous);
      }
      this.contentElement.classList.add(...next);
      this.outerElement.classList.add(...next);
    }, {
      fireImmediately: true
    })]);
  }
  componentWillUnmount() {
    this.props.layers.removeSheet(this);
    if (this.props.open && this.props.onCloseAnimationComplete) {
      this.closeTimeout = window.setTimeout(this.props.onCloseAnimationComplete, this.enableAnimations ? layer_view_1.ANIMATION_TIME : 0);
    }
  }
  componentDidUpdate(prevProps) {
    var _a, _b;
    const {
      open,
      content,
      outerContent,
      layers,
      placeholder = defaultPlaceholder,
      scrollable,
      onCloseAnimationComplete
    } = this.props;
    if (this.outerElement == null || this.contentElement == null || this.window == null) {
      return;
    }
    this.contentElement.classList.toggle(sheet_css_1.default.scrollable, scrollable);
    if (open && !prevProps.open) {
      this.show = true;
      this.contentElement.style.display = '';
      window.clearTimeout(this.closeTimeout);
      if (!this.currentContent) {
        this.currentContent = placeholder;
        this.currentOuterContent = outerContent;
        window.setTimeout((0, mobx_1.action)(() => {
          this.currentContent = content;
        }));
      }
      layers.addSheet(this);
    } else if (!open && prevProps.open) {
      this.show = false;
      (_b = (_a = this.sheetDragInternal) === null || _a === void 0 ? void 0 : _a.sheetBehavior) === null || _b === void 0 ? void 0 : _b.closeSheet();
      this.closeTimeout = window.setTimeout((0, mobx_1.action)(() => {
        if (this.unmountOnClose) {
          this.currentContent = undefined;
          this.currentOuterContent = undefined;
        }
        if (this.contentElement != null) {
          this.contentElement.style.display = 'none';
        }
        onCloseAnimationComplete && onCloseAnimationComplete();
      }), this.enableAnimations ? layer_view_1.ANIMATION_TIME : 0);
      this.unmountOnClose && layers.removeSheet(this);
    }
    if (this.contentShouldUpdate()) {
      this.currentContent = content;
      this.currentOuterContent = outerContent;
    }
  }
  contentShouldUpdate() {
    const {
      placeholder = defaultPlaceholder,
      mountBehaviour
    } = this.props;
    if (placeholder != null && this.currentContent === placeholder) {
      return false;
    }
    return this.show || mountBehaviour === 'eager';
  }
  get unmountOnClose() {
    return !this.props.mountBehaviour || this.props.mountBehaviour === 'open';
  }
  get applyInsetStyles() {
    const {
      expandContentIntoUnsafeArea = false,
      sheetContainsTextInput = false
    } = this.props;
    return !!this.supportsSafeAreaInset && !expandContentIntoUnsafeArea && !sheetContainsTextInput;
  }
  get from() {
    const {
      from,
      easelConfiguration: {
        direction
      }
    } = this.props;
    if (from === 'right' && direction === 'RTL') {
      return 'left';
    } else if (from === 'left' && direction === 'RTL') {
      return 'right';
    } else {
      return from || 'bottom';
    }
  }
  get onBottom() {
    return this.props.position === 'bottom';
  }
  get roundedCorners() {
    const {
      roundedCorners = true,
      largeRoundedCorners = false
    } = this.props;
    if (roundedCorners && largeRoundedCorners) {
      return 'large';
    }
    if (roundedCorners) {
      return 'small';
    }
    return 'none';
  }
  get autoFocusOnOpen() {
    const {
      overlay,
      open,
      autoFocusOnContent = true
    } = this.props;
    return !!(overlay && open && autoFocusOnContent);
  }
  get enableAnimations() {
    var _a;
    return ((_a = this.props.enableAnimations) !== null && _a !== void 0 ? _a : true) && this.props.easelConfiguration.enableAnimations;
  }
  get isFullscreen() {
    const {
      size,
      minSize
    } = this.props;
    if (!size) {
      return false;
    }
    const sheetSize = size || minSize;
    return sheetSize === '100%' || sheetSize === '100vh';
  }
  get contentStyles() {
    const {
      from,
      size,
      minSize,
      maxSize
    } = this.props;
    const nSize = (minSize != null || maxSize != null) && size == null ? 'inherit' : size;
    const nMaxSize = maxSize || '100%';
    return from === 'left' || from === 'right' ? {
      width: nSize,
      minWidth: minSize,
      maxWidth: nMaxSize
    } : {
      height: nSize,
      minHeight: minSize,
      maxHeight: nMaxSize
    };
  }
};
exports.InternalSheet = InternalSheet;
InternalSheet.defaultProps = {
  scrollable: true
};
InternalSheet.ids = new id_generator_1.IdGenerator('sheet__');
exports.InternalSheet = InternalSheet = InternalSheet_1 = tslib_1.__decorate([mobx_react_1.observer], InternalSheet);
