// @canva/app-ui-kit/dist/cjs/ui/base/box/box.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Box = void 0;
const classnames_1 = require("classnames");
const React = require("react");
const atoms_1 = require("./internal/atoms/atoms");
const types_1 = require("./types");
exports.Box = React.forwardRef(function Box(props, ref) {
    const { tagName = 'div', children, className, display, background, border, borderTop, borderBottom, borderStart, borderEnd, borderRadius, padding, paddingX, paddingY, paddingTop, paddingBottom, paddingStart, paddingEnd, width, height, minWidth, minHeight, flexDirection, flexWrap, justifyContent, alignItems, ...rest } = props;
    return React.createElement(tagName, {
        className: (0, classnames_1.default)(className, (0, atoms_1.resetAtom)(true), (0, atoms_1.backgroundAtom)(background), (0, atoms_1.borderAtom)(border), ...(0, atoms_1.borderPositionAtom)({
            top: borderTop,
            bottom: borderBottom,
            start: borderStart,
            end: borderEnd,
        }), (0, atoms_1.borderRadiusAtom)(borderRadius), (0, atoms_1.widthAtom)(width), (0, atoms_1.heightAtom)(height), (0, atoms_1.minWidthAtom)(minWidth), (0, atoms_1.minHeightAtom)(minHeight), ...(0, atoms_1.getDisplayClassNames)(display), ...(0, atoms_1.getPaddingClassNames)(padding, paddingX, paddingY, paddingTop, paddingBottom, paddingStart, paddingEnd), ...(0, atoms_1.getFlexDirectionClassNames)(flexDirection), (0, atoms_1.flexWrapAtom)(flexWrap), (0, atoms_1.justifyContentAtom)(justifyContent, display), (0, atoms_1.alignItemsAtom)(alignItems, display)),
        style: {
            ...(0, atoms_1.getDisplayStyles)(display),
            ...(0, atoms_1.getPaddingStyles)(padding, paddingX, paddingY, paddingTop, paddingBottom, paddingStart, paddingEnd),
            ...(0, atoms_1.getFlexDirectionStyles)(flexDirection),
        },
        ref,
        ...(0, types_1.commonPropsToDOMProps)(rest),
    }, children);
});
