// @canva/app-ui-kit/dist/cjs/ui/base/surface/flyout/flyout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Flyout = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const preconditions_1 = require('../../../../base/preconditions');
const React = require("react");
const device_capabilities_1 = require('../../device_capabilities/device_capabilities');
const popover_1 = require('../popover/popover');
const sheet_1 = require('../sheet/sheet');
const flyout_css_1 = require("./flyout.css");
function Flyout(
    { open, onRequestClose, onCloseComplete, trigger, width = '32u', dangerouslySetWidthMode = 'fixed', header, footer, children, ...contentContainerProps }
) {
    const mode = useMode();
    const resolveContent = (content) => typeof content === 'function' ? content({ mode }) : content;
    const resolvedHeader = resolveContent(header);
    const resolvedFooter = resolveContent(footer);
    const resolvedContent = ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [hasContent(resolvedHeader) && (0, jsx_runtime_1.jsx)("div", { className: flyout_css_1.default.header, children: resolvedHeader }), (0, jsx_runtime_1.jsx)("div", { className: flyout_css_1.default.content, children: resolveContent(children) }), hasContent(resolvedFooter) && (0, jsx_runtime_1.jsx)("div", { className: flyout_css_1.default.footer, children: resolvedFooter })] }));
    switch (mode) {
        case 'popover':
            const popoverWidth = width === 'trigger' ? 'reference' : width;
            return ((0, jsx_runtime_1.jsx)(popover_1.Popover, { open: open, onRequestClose: onRequestClose, onCloseComplete: onCloseComplete, reference: trigger, width: popoverWidth, widthMode: dangerouslySetWidthMode, ...contentContainerProps, children: resolvedContent }));
        case 'sheet':
            return ((0, jsx_runtime_1.jsx)(FlyoutSheet, { open: open, onRequestClose: onRequestClose, onCloseComplete: onCloseComplete, trigger: trigger, ...contentContainerProps, children: resolvedContent }));
        default:
            throw new preconditions_1.UnreachableError(mode);
    }
}
exports.Flyout = Flyout;
function useMode() {
    return (0, device_capabilities_1.isPortraitMobileScreen)() ? 'sheet' : 'popover';
}
const hasContent = (node) => React.Children.count(node) > 0;
function FlyoutSheet({ id: idProp, open, onRequestClose, onCloseComplete, trigger, handleSafeAreaInsetBottom = true, children, ...contentContainerProps }) {
    const defaultId = React.useId();
    const id = idProp || defaultId;
    return ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [typeof trigger === 'function'
                ? trigger({
                    ariaControls: open ? id : undefined,
                    active: open,
                    disclosure: true,
                    ariaHasPopup: contentContainerProps.role,
                })
                : !(0, popover_1.isReferenceObject)(trigger) && trigger, (0, jsx_runtime_1.jsx)(sheet_1.Sheet, { open: open, onRequestClose: onRequestClose, onCloseComplete: onCloseComplete, id: id, handleSafeAreaInsetBottom: handleSafeAreaInsetBottom, ...contentContainerProps, children: children })] }));
}
