// @canva/app-ui-kit/dist/cjs/ui/base/tooltip/tooltip_presenter.js
"use strict";

const makeObservable = require('../../../base/make_observable/make_observable').makeObservable;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TooltipPresenter = exports.TooltipState = void 0;
const tslib_1 = require("tslib");
const mobx = require("mobx");
class TooltipState {
  static _makeObservable(instance) {
    makeObservable(instance, {
      open: mobx.observable.ref
    });
  }
  constructor({
    open = false
  } = {}) {
    TooltipState._makeObservable(this);
    this.timeoutId = 0;
    this.open = open;
  }
}
exports.TooltipState = TooltipState;
class TooltipPresenter {
  static _makeObservable(instance) {
    makeObservable(instance, {
      setOpen: mobx.action
    });
  }
  constructor() {
    TooltipPresenter._makeObservable(this);
  }
  showTooltip(tooltip, fadeInDelayMs) {
    this.clearDelay(tooltip);
    return TooltipPresenter.fadeWithDelay(tooltip, fadeInDelayMs, () => this.setOpen(tooltip, true));
  }
  hideTooltip(tooltip, fadeOutDelayMs) {
    this.clearDelay(tooltip);
    return TooltipPresenter.fadeWithDelay(tooltip, fadeOutDelayMs, () => this.setOpen(tooltip, false));
  }
  setOpen({
    state
  }, open) {
    state.open = open;
  }
  static fadeWithDelay({
    state
  }, delay, cb) {
    if (!delay) {
      return Promise.resolve(cb());
    }
    return new Promise(res => {
      state.timeoutId = window.setTimeout(() => res(cb()), delay);
    });
  }
  clearDelay({
    state
  }) {
    window.clearTimeout(state.timeoutId);
  }
}
exports.TooltipPresenter = TooltipPresenter;
