// @canva/app-ui-kit/dist/cjs/ui/base/form/input/base_input.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseInput = exports.useOnWrapperMouseDown = exports.useFocusState = exports.useInputValue = exports.useInputControls = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const compose_react_refs_1 = require("@seznam/compose-react-refs");
const exists_1 = require('../../../../base/exists');
const preconditions_1 = require('../../../../base/preconditions');
const classnames_1 = require("classnames");
const mobx_1 = require("mobx");
const mobx_react_1 = require("mobx-react");
const React = require("react");
const access_mode_state_1 = require('../../a11y/access_mode_state/access_mode_state');
const controllable_value_1 = require('../../controllable_value/controllable_value');
const device_capabilities_1 = require('../../device_capabilities/device_capabilities');
const icon_1 = require('../../icons/calendar/icon');
const icon_2 = require('../../icons/credit_card_number/icon');
const icon_3 = require('../../icons/credit_card_verification/icon');
const icon_4 = require('../../icons/email/icon');
const icon_5 = require('../../icons/link/icon');
const icon_6 = require('../../icons/location/icon');
const icon_7 = require('../../icons/person/icon');
const icon_8 = require('../../icons/search/icon');
const spacer_1 = require('../../layout/spacer/spacer');
const base_input_css_1 = require("./base_input.css");
const input_internal_1 = require("./internal/input_internal");
var input_internal_2 = require("./internal/input_internal");
Object.defineProperty(exports, "useInputControls", { enumerable: true, get: function () { return input_internal_2.useInputControls; } });
Object.defineProperty(exports, "useInputValue", { enumerable: true, get: function () { return input_internal_2.useInputValue; } });
function getLegacyIcon({ icon, iconClassName }, position) {
    return (icon === null || icon === void 0 ? void 0 : icon.align) === position && (0, jsx_runtime_1.jsx)(Icon, { ...icon, className: iconClassName });
}
function getDecoration(decoration) {
    const content = typeof decoration === 'function' ? decoration() : decoration;
    return content
        ? ((0, jsx_runtime_1.jsx)("div", { className: base_input_css_1.default.decoration, children: React.isValidElement(content) && content.type === React.Fragment
                ? React.Children.toArray(content.props.children).filter(exists_1.exists).map(addLeadingSpace)
                : content }))
        : undefined;
}
function addLeadingSpace(child, index) {
    return index === 0 ? child : [(0, jsx_runtime_1.jsx)(spacer_1.Spacer, { size: "0.5u" }, index), child];
}
function useFocusState() {
    const store = (0, mobx_react_1.useLocalObservable)(() => ({
        isFocused: false,
        setFocused: (0, mobx_1.action)((value) => (store.isFocused = value)),
    }));
    return {
        isFocused: store.isFocused,
        isKeyboardMode: access_mode_state_1.accessModeState.isKeyboardMode,
        setFocused: store.setFocused,
    };
}
exports.useFocusState = useFocusState;
function useOnWrapperMouseDown() {
    const ref = React.useRef(null);
    const onMouseDown = React.useCallback((e) => {
        var _a;
        let el = e.target;
        while (el && el !== e.currentTarget) {
            if (isInteractive(el)) {
                return;
            }
            el = el.parentElement;
        }
        e.stopPropagation();
        e.preventDefault();
        (_a = ref.current) === null || _a === void 0 ? void 0 : _a.focus();
    }, []);
    return { ref, onMouseDown };
}
exports.useOnWrapperMouseDown = useOnWrapperMouseDown;
function isInteractive(el) {
    return (el.tagName === 'A'
        || el.tagName === 'BUTTON'
        || el.tagName === 'INPUT'
        || el.tagName === 'TEXTAREA'
        || el.getAttribute('tabIndex') != null);
}
exports.BaseInput = (0, mobx_react_1.observer)(React.forwardRef(InternalBaseInput));
function InternalBaseInput(props_, forwardedRef) {
    var _a, _b, _c;
    const { props, setPropOverrides } = (0, input_internal_1.usePropOverrides)(props_);
    const { blurOnEnterKeyDown, onChange, onFocus, onBlur, onKeyDown, onChangeComplete, } = props;
    const { ref: inputRef, onMouseDown: onWrapperMouseDown } = useOnWrapperMouseDown();
    const { isFocused, isKeyboardMode, setFocused } = useFocusState();
    const [value, setValue] = (0, controllable_value_1.useControllableValue)({
        value: props.value,
        defaultValue: (_a = props.defaultValue) !== null && _a !== void 0 ? _a : '',
    });
    const startDecoration = getDecoration((_b = props.start) !== null && _b !== void 0 ? _b : getLegacyIcon(props, 'start'));
    const endDecoration = getDecoration((_c = props.end) !== null && _c !== void 0 ? _c : getLegacyIcon(props, 'end'));
    const controls = (0, input_internal_1.useCreateInputControls)({
        focus: () => { var _a; return (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus(); },
        setType: type => setPropOverrides(type ? { type } : undefined),
        setDisabled: disabled => setPropOverrides({ disabled }),
        setValue: (value) => {
            setValue(value);
            onChange === null || onChange === void 0 ? void 0 : onChange(value);
        },
    });
    const _onChange = React.useCallback((e) => {
        setValue(e.target.value);
        onChange === null || onChange === void 0 ? void 0 : onChange(e.target.value, e);
    }, [onChange, setValue]);
    const _onFocus = React.useCallback((e) => {
        onFocus === null || onFocus === void 0 ? void 0 : onFocus(e);
        setFocused(true);
    }, [onFocus, setFocused]);
    const _onBlur = React.useCallback((e) => {
        onBlur === null || onBlur === void 0 ? void 0 : onBlur(e);
        onChangeComplete === null || onChangeComplete === void 0 ? void 0 : onChangeComplete(e.target.value);
        setFocused(false);
    }, [onBlur, onChangeComplete, setFocused]);
    const _onKeyDown = React.useCallback((e) => {
        var _a;
        if (e.keyCode === 229) {
            e.stopPropagation();
            return;
        }
        if (e.key === 'Enter' && blurOnEnterKeyDown) {
            (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.blur();
        }
        onKeyDown === null || onKeyDown === void 0 ? void 0 : onKeyDown(e);
        if (!blurOnEnterKeyDown && e.key === 'Enter') {
            onChangeComplete === null || onChangeComplete === void 0 ? void 0 : onChangeComplete(e.currentTarget.value);
        }
    }, [onKeyDown, onChangeComplete, blurOnEnterKeyDown, inputRef]);
    const wrapperClassName = (0, classnames_1.default)(base_input_css_1.default.wrapper, {
        [base_input_css_1.default.hoverSupported]: (0, device_capabilities_1.canHover)(),
        [base_input_css_1.default.focusOutline]: isFocused && isKeyboardMode,
        [base_input_css_1.default.borderless]: props.borderless,
        [base_input_css_1.default.withStartDecoration]: startDecoration != null,
        [base_input_css_1.default.withEndDecoration]: endDecoration != null,
    }, 
    (props.disabled && base_input_css_1.default.disabled)
        || (isFocused && base_input_css_1.default.active) || (props.error && base_input_css_1.default.error), props.className);
    const inputClassName = (0, classnames_1.default)(base_input_css_1.default.textField, {
        [base_input_css_1.default.textAlignCenter]: props.textAlignCenter,
        [base_input_css_1.default.noAutocomplete]: props.autoComplete === 'off',
    }, props.inputClassName);
    return ((0, jsx_runtime_1.jsx)(
        "div",
        { className: wrapperClassName, onMouseDown: onWrapperMouseDown, children: (0, jsx_runtime_1.jsxs)(input_internal_1.BaseInputContextProvider, { controls: controls, value: value, children: [startDecoration, (0, jsx_runtime_1.jsx)("input", { id: props.id, className: inputClassName, value: value, onChange: _onChange, onFocus: _onFocus, onBlur: _onBlur, onKeyDown: _onKeyDown, onClick: props.onClick, onMouseDown: props.onMouseDown, onMouseUp: props.onMouseUp, onContextMenu: props.onContextMenu, onPaste: props.onPaste, onKeyUp: props.onKeyUp, ref: (0, compose_react_refs_1.default)(inputRef, forwardedRef), dir: "auto", spellCheck: props.disableSpellcheck ? false : undefined, type: props.type, inputMode: props.inputMode, autoCapitalize: props.autoCapitalize, autoCorrect: props.autoCorrect, pattern: props.pattern, autoFocus: props.autoFocus === 'always'
                            || (props.autoFocus === 'on-desktop' && !(0, device_capabilities_1.isVirtualKeyboard)()), autoComplete: props.autoComplete, disabled: props.disabled, readOnly: props.readOnly, maxLength: props.maxLength, max: props.max, min: props.min, placeholder: props.placeholder, step: props.step, name: props.name, role: props.role, "aria-roledescription": props.ariaRoleDescription, "aria-required": props.required || undefined, "aria-invalid": props.error || undefined, "aria-label": props.ariaLabel, "aria-labelledby": props.ariaLabelledBy, "aria-describedby": props.ariaDescribedBy, "aria-autocomplete": props.ariaAutoComplete, "aria-activedescendant": props.ariaActiveDescendant, "aria-controls": props.ariaControls, "aria-expanded": props.ariaExpanded, "aria-haspopup": props.ariaHasPopup, "aria-valuenow": props.ariaValueNow, "aria-valuetext": props.ariaValueText, "aria-valuemin": props.ariaValueMin, "aria-valuemax": props.ariaValueMax }), endDecoration] }) }
    ));
}
function Icon({ type, className }) {
    const props = {
        size: 'medium',
        tone: 'primary',
        className,
    };
    switch (type) {
        case 'credit-card':
            return (0, jsx_runtime_1.jsx)(icon_2.CreditCardNumberIcon, { ...props });
        case 'credit-card-expiry':
            return (0, jsx_runtime_1.jsx)(icon_1.CalendarIcon, { ...props });
        case 'credit-card-verification':
            return (0, jsx_runtime_1.jsx)(icon_3.CreditCardVerificationIcon, { ...props });
        case 'email':
            return (0, jsx_runtime_1.jsx)(icon_4.EmailIcon, { ...props });
        case 'link':
            return (0, jsx_runtime_1.jsx)(icon_5.LinkIcon, { ...props });
        case 'location':
            return (0, jsx_runtime_1.jsx)(icon_6.LocationIcon, { ...props });
        case 'person':
            return (0, jsx_runtime_1.jsx)(icon_7.PersonIcon, { ...props });
        case 'search':
            return (0, jsx_runtime_1.jsx)(icon_8.SearchIcon, { ...props });
        default:
            throw new preconditions_1.UnreachableError(type);
    }
}
