// @canva/app-ui-kit/dist/cjs/ui/base/form/input/number_input.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.snapToStep = exports.getRoundedNumberValue = exports.getFractionDigits = exports.isValidNumericKeyboardEvent = exports.isDecimalCharacter = exports.NumberInput = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const compose_react_refs_1 = require("@seznam/compose-react-refs");
const React = require("react");
const button_1 = require('../../button/button');
const controllable_value_1 = require('../../controllable_value/controllable_value');
const icon_1 = require('../../icons/minus/icon');
const icon_2 = require('../../icons/plus/icon');
const base_input_1 = require("./base_input");
exports.NumberInput = React.forwardRef(function NumberInput({ value: valueProp, defaultValue, onChange: onChangeProp, onChangeComplete: onChangeCompleteProp, onKeyDown: onKeyDownProp, step = 1, min, max, maximumFractionDigits, isValidCharacter: isValidCharacterProp, hasSpinButtons, decrementAriaLabel, incrementAriaLabel, ...passThroughProps }, ref) {
    const [value, setValue] = (0, controllable_value_1.useControllableValue)({
        value: valueProp != null ? valueProp.toString() : undefined,
        defaultValue: defaultValue != null ? defaultValue.toString() : undefined,
    });
    const hasDecimals = maximumFractionDigits == null || maximumFractionDigits > 0;
    const hasNegative = min == null || min < 0;
    const [inputMode, setInputMode] = React.useState(hasDecimals ? 'decimal' : 'numeric');
    React.useEffect(() => {
        if (isIphone() && hasNegative) {
            setInputMode('text');
        }
        else if (isAndroid() && hasNegative) {
            setInputMode('numeric');
        }
    }, [hasNegative]);
    const isValidCharacter = React.useMemo(() => {
        if (isValidCharacterProp) {
            return isValidCharacterProp;
        }
        if (hasDecimals) {
            return hasNegative ? exports.isDecimalCharacter : isNonNegativeDecimalCharacter;
        }
        return hasNegative ? isIntegerCharacter : isNonNegativeIntegerCharacter;
    }, [isValidCharacterProp, hasDecimals, hasNegative]);
    const updateValue = React.useCallback((newNumberValue, newValue) => {
        if (newValue !== value) {
            setValue(newValue);
            onChangeProp && onChangeProp(newNumberValue, newValue);
        }
    }, [onChangeProp, value, setValue]);
    const commit = React.useCallback((finalNumberValue) => {
        const stringValue = finalNumberValue == null ? '' : finalNumberValue.toString();
        updateValue(finalNumberValue, stringValue);
        onChangeCompleteProp && onChangeCompleteProp(finalNumberValue);
    }, [onChangeCompleteProp, updateValue]);
    const castedValue = cast(value || '');
    const canChange = !passThroughProps.disabled && !passThroughProps.readOnly;
    const canStep = canChange && (value == null || value === '' || Number.isFinite(castedValue));
    const snapStep = step === 'any' ? 1 : step;
    const getNextStep = React.useCallback((operation) => {
        if (value == null || value === '') {
            const minOrMax = operation === '+' ? max : min;
            return minOrMax != null ? minOrMax : snapStep;
        }
        const currentNumberValue = castedValue;
        const snappedValue = (0, exports.snapToStep)(currentNumberValue, snapStep, min, max);
        if ((operation === '+' && snappedValue > currentNumberValue)
            || (operation === '-' && snappedValue < currentNumberValue)) {
            return snappedValue;
        }
        if (operation === '+') {
            return (0, exports.snapToStep)(snappedValue + snapStep, snapStep, min, max);
        }
        return (0, exports.snapToStep)(snappedValue - snapStep, snapStep, min, max);
    }, [value, castedValue, min, max, snapStep]);
    const increment = React.useCallback(() => {
        if (!canStep) {
            return;
        }
        const newValue = getNextStep('+');
        commit(newValue);
    }, [canStep, commit, getNextStep]);
    const decrement = React.useCallback(() => {
        if (!canStep) {
            return;
        }
        const newValue = getNextStep('-');
        commit(newValue);
    }, [canStep, commit, getNextStep]);
    const setToMin = React.useCallback(() => {
        if (!canChange || min == null) {
            return;
        }
        commit(min);
    }, [canChange, min, commit]);
    const setToMax = React.useCallback(() => {
        if (!canChange || max == null) {
            return;
        }
        commit(max);
    }, [canChange, max, commit]);
    const onChange = React.useCallback((text) => {
        if (!canChange) {
            return;
        }
        const newValue = sanitize(text, isValidCharacter);
        updateValue(cast(newValue), newValue);
    }, [canChange, updateValue, isValidCharacter]);
    const onChangeComplete = React.useCallback((text) => {
        if (!canChange) {
            return;
        }
        const finalNumberValue = getFinalNumberValue(text, min, max, maximumFractionDigits);
        commit(finalNumberValue);
    }, [canChange, commit, min, max, maximumFractionDigits]);
    const inputRef = React.createRef();
    const onKeyDown = React.useCallback((e) => {
        var _a;
        const selectionStart = ((_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.selectionStart) || undefined;
        if (!(0, exports.isValidNumericKeyboardEvent)(e, isValidCharacter, value, selectionStart)) {
            e.preventDefault();
            return;
        }
        onKeyDownProp && onKeyDownProp(e);
        switch (e.key) {
            case 'ArrowUp':
                increment();
                break;
            case 'ArrowDown':
                decrement();
                break;
            case 'Home':
                e.preventDefault();
                setToMin();
                break;
            case 'End':
                e.preventDefault();
                setToMax();
                break;
            default:
                break;
        }
    }, [onKeyDownProp, isValidCharacter, value, increment, decrement, setToMax, setToMin, inputRef]);
    const start = hasSpinButtons
        ? ((0, jsx_runtime_1.jsx)(SpinButton, { operation: "-", onClick: decrement, ariaLabel: decrementAriaLabel, disabled: !canChange }))
        : undefined;
    const end = hasSpinButtons
        ? ((0, jsx_runtime_1.jsx)(SpinButton, { operation: "+", onClick: increment, ariaLabel: incrementAriaLabel, disabled: !canChange }))
        : undefined;
    const spinButtonProps = isIOS()
        ? {}
        : {
            role: 'spinbutton',
            ariaValueNow: Number.isFinite(castedValue) ? castedValue : undefined,
            ariaValueText: value,
            ariaValueMin: min,
            ariaValueMax: max,
        };
    return ((0, jsx_runtime_1.jsx)(base_input_1.BaseInput, { ref: (0, compose_react_refs_1.default)(ref, inputRef), type: "text", inputMode: inputMode, value: value == null ? '' : value, onChange: onChange, onChangeComplete: onChangeComplete, onKeyDown: onKeyDown, start: start, end: end, textAlignCenter: hasSpinButtons ? true : undefined, ...spinButtonProps, ...passThroughProps }));
});
exports.NumberInput.displayName = 'NumberInput';
const SpinButton = ({ onClick, operation, disabled, ariaLabel }) => {
    return ((0, jsx_runtime_1.jsx)(button_1.Button, { variant: "tertiary", size: "small", onClick: onClick, repeatSpeed: "slow", icon: operation === '+' ? icon_2.PlusIcon : icon_1.MinusIcon, ariaLabel: ariaLabel, tabIndex: -1, disabled: disabled }));
};
const isIOS = () => {
    return navigator != null && !!navigator.userAgent.match(/iphone|ipad/gi);
};
const isIphone = () => {
    return navigator != null && !!navigator.userAgent.match(/iphone/gi);
};
const isAndroid = () => {
    return navigator != null && !!navigator.userAgent.match(/android/gi);
};
const isNonNegativeIntegerCharacter = (character) => {
    return /^[0-9]$/.test(character);
};
const isIntegerCharacter = (character) => {
    return /^[0-9-]$/.test(character);
};
const isNonNegativeDecimalCharacter = (character) => {
    return /^[0-9.]$/.test(character);
};
const isDecimalCharacter = (character) => {
    return /^[0-9\-.]$/.test(character);
};
exports.isDecimalCharacter = isDecimalCharacter;
const isValidNumericKeyboardEvent = (e, isValidCharacter, value, 
selectionStart = 0) => {
    if (e.key == null || e.ctrlKey || e.altKey || e.metaKey) {
        return true;
    }
    if (e.key.length !== 1) {
        return true;
    }
    if (!isValidCharacter(e.key)) {
        return false;
    }
    if (e.key === '-') {
        return selectionStart === 0;
    }
    if (e.key === '.') {
        return selectionStart !== 0 && value != null && !value.includes('.');
    }
    return true;
};
exports.isValidNumericKeyboardEvent = isValidNumericKeyboardEvent;
const sanitize = (text, isValidCharacter) => {
    return text.split('').filter(isValidCharacter).join('');
};
const cast = (text) => {
    const value = text === '' ? undefined : Number(text);
    return value;
};
const clamp = (value, min, max) => {
    if (min != null && min > value) {
        return min;
    }
    if (max != null && max < value) {
        return max;
    }
    return value;
};
const getFractionDigits = (value) => {
    const stringValue = value.toString();
    const index = stringValue.indexOf('.');
    return index === -1 ? 0 : stringValue.length - index - 1;
};
exports.getFractionDigits = getFractionDigits;
const roundToPrecision = (value, maximumFractionDigits) => {
    const pow = Math.pow(10, maximumFractionDigits);
    return Math.round(value * pow) / pow;
};
const getRoundedNumberValue = (value, maximumFractionDigits) => {
    if (value == null) {
        return undefined;
    }
    const numberValue = typeof value === 'string' ? cast(value) : value;
    return numberValue == null ? undefined : roundToPrecision(numberValue, maximumFractionDigits);
};
exports.getRoundedNumberValue = getRoundedNumberValue;
const getFinalNumberValue = (text, min, max, maximumFractionDigits) => {
    const value = cast(text);
    if (Number.isNaN(value) || value == null) {
        return undefined;
    }
    let result = clamp(value, min, max);
    if (maximumFractionDigits != null) {
        result = roundToPrecision(result, maximumFractionDigits);
    }
    return result;
};
const snapToStep = (value, step, min, max) => {
    const remainder = (value - (min == null ? 0 : min)) % step;
    let result = Math.abs(remainder) * 2 >= step
        ? value + Math.sign(remainder) * (step - Math.abs(remainder))
        : value - remainder;
    if (min != null) {
        if (result < min) {
            result = min;
        }
        else if (max != null && result > max) {
            result = min + Math.floor((max - min) / step) * step;
        }
    }
    else if (max != null && result > max) {
        result = Math.floor(max / step) * step;
    }
    const precision = (0, exports.getFractionDigits)(step);
    if (precision > 0) {
        result = roundToPrecision(result, precision);
    }
    return result;
};
exports.snapToStep = snapToStep;
