// @canva/app-ui-kit/dist/cjs/ui/base/layout/rows/rows.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rows = void 0;
const classnames_1 = require("classnames");
const React = require("react");
const types_1 = require('../../box/types');
const metrics_1 = require('../../metrics/metrics');
const responsive_1 = require('../../responsive/responsive');
const rows_css_1 = require("./rows.css");
const spacingPropertyMap = {
    default: rows_css_1.customProperties.spacing,
    smallUp: rows_css_1.customProperties.smallUpSpacing,
    mediumUp: rows_css_1.customProperties.mediumUpSpacing,
    largeUp: rows_css_1.customProperties.largeUpSpacing,
};
const alignPropertyMap = {
    default: rows_css_1.customProperties.align,
    smallUp: rows_css_1.customProperties.smallUpAlign,
    mediumUp: rows_css_1.customProperties.mediumUpAlign,
    largeUp: rows_css_1.customProperties.largeUpAlign,
};
exports.Rows = React.forwardRef(function Rows({ children, spacing, align, tagName = 'div', ...rest }, ref) {
    return React.createElement(tagName, {
        className: (0, classnames_1.default)(rows_css_1.default.rows),
        style: {
            ...(spacing && (0, responsive_1.generateStyle)(spacingPropertyMap, spacing, metrics_1.getSpaceValue)),
            ...(align && (0, responsive_1.generateStyle)(alignPropertyMap, align)),
        },
        children,
        ref,
        ...(0, types_1.commonPropsToDOMProps)(rest),
    });
});
exports.Rows.displayName = 'Rows';
