// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/initialize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mobx_1 = require("mobx");
const key_1 = require('../../../base/bootstrap/key');
const ui_bootstrap_proto_1 = require('../../../base/bootstrap/ui_bootstrap_proto');
initializeEaselBootstrap({
    enableAnimations: true,
    direction: ui_bootstrap_proto_1.Direction.LTR,
    disableFocusTraps: true,
    enableUserSelection: true,
    disableDialogBlur: false,
    exposeStringIds: false,
    enableHighColorContrast: false,
    autoplayVideos: undefined,
    theme: ui_bootstrap_proto_1.Theme.DARK,
    enableEaselUiRefresh: false,
});
(0, mobx_1.configure)({
    enforceActions: 'observed',
    isolateGlobalState: true,
});
function initializeEaselBootstrap(bootstrap) {
    document.documentElement.dir = bootstrap.direction === ui_bootstrap_proto_1.Direction.RTL ? 'rtl' : 'ltr';
    window['bootstrap'] = {
        [key_1.uiBootstrapKey]: ui_bootstrap_proto_1.UiBootstrap.serialize({
            ...bootstrap,
            direction: bootstrap.direction,
        }),
    };
}
