// @canva/app-ui-kit/dist/cjs/ui/base/theme/internal.js
"use strict";

const makeObservable = require('../../../base/make_observable/make_observable').makeObservable;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ThemeContext = exports.rootThemeStore = exports.ThemeData = exports.ThemeStore = exports.getAllPreloadClasses = exports.getThemeClasses = exports.stripGlobal = void 0;
const tslib_1 = require("tslib");
const preconditions_1 = require('../../../base/preconditions');
const classnames_1 = require("classnames");
const mobx_1 = require("mobx");
const React = require("react");
const preload_css_1 = require("./preload.css");
const theme_css_1 = require("./theme.css");
const stripGlobal = className => {
  var _a;
  return (_a = className === null || className === void 0 ? void 0 : className.replace(/^\s*:global\(\.(.+?)\)\s*$/, '$1')) !== null && _a !== void 0 ? _a : '';
};
exports.stripGlobal = stripGlobal;
let themeClasses;
let themedClass;
function getThemeClassMap() {
  if (themeClasses == null) {
    themeClasses = new Map([['light', (0, exports.stripGlobal)(theme_css_1.default.themeLightClass)], ['dark', (0, exports.stripGlobal)(theme_css_1.default.themeDark)]]);
  }
  return themeClasses;
}
function getThemedClass() {
  if (themedClass == null) {
    themedClass = (0, exports.stripGlobal)(theme_css_1.default.themed);
  }
  return themedClass;
}
function getThemeClasses(theme) {
  const themeClasses = getThemeClassMap();
  return [getThemedClass(), preconditions_1.Preconditions.checkExists(themeClasses.get(theme))];
}
exports.getThemeClasses = getThemeClasses;
function getAllPreloadClasses() {
  return [getThemedClass(), ...getThemeClassMap().values(), (0, exports.stripGlobal)(preload_css_1.default.themePreloadModeClassic)];
}
exports.getAllPreloadClasses = getAllPreloadClasses;
class ThemeStore {
  static _makeObservable(instance) {
    makeObservable(instance, {
      parent: mobx_1.observable.ref,
      _appearance: mobx_1.observable.ref,
      themeMapping: mobx_1.observable.deep,
      setParent: mobx_1.action,
      setAppearance: mobx_1.action,
      appearance: mobx_1.computed,
      setThemeMapping: mobx_1.action,
      currentTheme: mobx_1.computed
    });
  }
  constructor() {
    ThemeStore._makeObservable(this);
    this.themeMapping = {};
  }
  setParent(store) {
    this.parent = store;
  }
  setAppearance(appearance) {
    this._appearance = appearance;
  }
  get appearance() {
    var _a, _b;
    return (_a = this._appearance) !== null && _a !== void 0 ? _a : (_b = this.parent) === null || _b === void 0 ? void 0 : _b.appearance;
  }
  setThemeMapping(themeMapping) {
    this.themeMapping = themeMapping;
  }
  get currentTheme() {
    var _a, _b;
    const appearance = this.appearance;
    if (appearance == null) {
      return undefined;
    }
    return (_a = this.themeMapping[appearance]) !== null && _a !== void 0 ? _a : (_b = this.parent) === null || _b === void 0 ? void 0 : _b.currentTheme;
  }
}
exports.ThemeStore = ThemeStore;
class ThemeData {
  static _makeObservable(instance) {
    makeObservable(instance, {
      classNames: mobx_1.computed,
      className: mobx_1.computed,
      mode: mobx_1.computed
    });
  }
  constructor(store) {
    ThemeData._makeObservable(this);
    this.store = store;
  }
  get classNames() {
    const {
      currentTheme
    } = this.store;
    if (currentTheme == null) {
      return [];
    }
    return getThemeClasses(currentTheme);
  }
  get className() {
    const classes = this.classNames;
    if (classes.length === 0) {
      return undefined;
    }
    return (0, classnames_1.default)(classes);
  }
  get mode() {
    switch (this.store.appearance) {
      case 'dark':
      case 'light':
        return 'modern';
      case 'classicDark':
      case 'classicLight':
        return 'classic';
      case undefined:
        return undefined;
      default:
        throw new preconditions_1.UnreachableError(this.store.appearance);
    }
  }
}
exports.ThemeData = ThemeData;
exports.rootThemeStore = new ThemeStore();
const rootThemeData = new ThemeData(exports.rootThemeStore);
exports.ThemeContext = React.createContext({
  store: exports.rootThemeStore,
  data: rootThemeData
});
