// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/styles/tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tokens = void 0;
const tokens_css_1 = require("./tokens.css");
exports.tokens = {
    space0: tokens_css_1.default.uiKitSpace0,
    space050: tokens_css_1.default.uiKitSpace050,
    space1: tokens_css_1.default.uiKitSpace1,
    space150: tokens_css_1.default.uiKitSpace150,
    space2: tokens_css_1.default.uiKitSpace2,
    space3: tokens_css_1.default.uiKitSpace3,
    space4: tokens_css_1.default.uiKitSpace4,
    space6: tokens_css_1.default.uiKitSpace6,
    space8: tokens_css_1.default.uiKitSpace8,
    space12: tokens_css_1.default.uiKitSpace12,
    baseUnit: tokens_css_1.default.uiKitBaseUnit,
    borderRadius: tokens_css_1.default.uiKitBorderRadius,
    minTouchableArea: tokens_css_1.default.uiKitMinTouchableArea,
    boxShadowFaint: tokens_css_1.default.uiKitBoxShadowFaint,
    boxShadowMedium: tokens_css_1.default.uiKitBoxShadowMedium,
    boxShadowHeavy: tokens_css_1.default.uiKitBoxShadowHeavy,
    hoverTransition: tokens_css_1.default.uiKitHoverTransition,
    fadeTransition: tokens_css_1.default.uiKitFadeTransition,
    colorOverlay: tokens_css_1.default.uiKitColorOverlay,
    colorPrimary: tokens_css_1.default.uiKitColorPrimary,
    colorPrimaryHover: tokens_css_1.default.uiKitColorPrimaryHover,
    colorPrimaryActive: tokens_css_1.default.uiKitColorPrimaryActive,
    colorPrimaryDisabled: tokens_css_1.default.uiKitColorPrimaryDisabled,
    colorPrimaryFore: tokens_css_1.default.uiKitColorPrimaryFore,
    colorPrimaryForeDisabled: tokens_css_1.default.uiKitColorPrimaryForeDisabled,
    colorSecondary: tokens_css_1.default.uiKitColorSecondary,
    colorSecondaryHover: tokens_css_1.default.uiKitColorSecondaryHover,
    colorSecondaryActive: tokens_css_1.default.uiKitColorSecondaryActive,
    colorSecondaryDisabled: tokens_css_1.default.uiKitColorSecondaryDisabled,
    colorSecondaryFore: tokens_css_1.default.uiKitColorSecondaryFore,
    colorSecondaryForeDisabled: tokens_css_1.default.uiKitColorSecondaryForeDisabled,
    colorTertiary: tokens_css_1.default.uiKitColorTertiary,
    colorTertiaryHover: tokens_css_1.default.uiKitColorTertiaryHover,
    colorTertiaryActive: tokens_css_1.default.uiKitColorTertiaryActive,
    colorTertiaryDisabled: tokens_css_1.default.uiKitColorTertiaryDisabled,
    colorTertiaryFore: tokens_css_1.default.uiKitColorTertiaryFore,
    colorTertiaryForeDisabled: tokens_css_1.default.uiKitColorTertiaryForeDisabled,
    colorContrast: tokens_css_1.default.uiKitColorContrast,
    colorContrastHover: tokens_css_1.default.uiKitColorContrastHover,
    colorContrastActive: tokens_css_1.default.uiKitColorContrastActive,
    colorContrastDisabled: tokens_css_1.default.uiKitColorContrastDisabled,
    colorContrastFore: tokens_css_1.default.uiKitColorContrastFore,
    colorContrastForeDisabled: tokens_css_1.default.uiKitColorContrastForeDisabled,
    colorNeutral: tokens_css_1.default.uiKitColorNeutral,
    colorNeutralHover: tokens_css_1.default.uiKitColorNeutralHover,
    colorNeutralActive: tokens_css_1.default.uiKitColorNeutralActive,
    colorNeutralDisabled: tokens_css_1.default.uiKitColorNeutralDisabled,
    colorNeutralLow: tokens_css_1.default.uiKitColorNeutralLow,
    colorNeutralFore: tokens_css_1.default.uiKitColorNeutralFore,
    colorNeutralForeLow: tokens_css_1.default.uiKitColorNeutralForeLow,
    colorNeutralForeDisabled: tokens_css_1.default.uiKitColorNeutralForeDisabled,
    colorPositive: tokens_css_1.default.uiKitColorPositive,
    colorPositiveHover: tokens_css_1.default.uiKitColorPositiveHover,
    colorPositiveActive: tokens_css_1.default.uiKitColorPositiveActive,
    colorPositiveDisabled: tokens_css_1.default.uiKitColorPositiveDisabled,
    colorPositiveLow: tokens_css_1.default.uiKitColorPositiveLow,
    colorPositiveFore: tokens_css_1.default.uiKitColorPositiveFore,
    colorPositiveForeLow: tokens_css_1.default.uiKitColorPositiveForeLow,
    colorPositiveForeDisabled: tokens_css_1.default.uiKitColorPositiveForeDisabled,
    colorInfo: tokens_css_1.default.uiKitColorInfo,
    colorInfoHover: tokens_css_1.default.uiKitColorInfoHover,
    colorInfoActive: tokens_css_1.default.uiKitColorInfoActive,
    colorInfoDisabled: tokens_css_1.default.uiKitColorInfoDisabled,
    colorInfoLow: tokens_css_1.default.uiKitColorInfoLow,
    colorInfoFore: tokens_css_1.default.uiKitColorInfoFore,
    colorInfoForeLow: tokens_css_1.default.uiKitColorInfoForeLow,
    colorInfoForeDisabled: tokens_css_1.default.uiKitColorInfoForeDisabled,
    colorWarn: tokens_css_1.default.uiKitColorWarn,
    colorWarnHover: tokens_css_1.default.uiKitColorWarnHover,
    colorWarnActive: tokens_css_1.default.uiKitColorWarnActive,
    colorWarnDisabled: tokens_css_1.default.uiKitColorWarnDisabled,
    colorWarnLow: tokens_css_1.default.uiKitColorWarnLow,
    colorWarnFore: tokens_css_1.default.uiKitColorWarnFore,
    colorWarnForeLow: tokens_css_1.default.uiKitColorWarnForeLow,
    colorWarnForeDisabled: tokens_css_1.default.uiKitColorWarnForeDisabled,
    colorCritical: tokens_css_1.default.uiKitColorCritical,
    colorCriticalHover: tokens_css_1.default.uiKitColorCriticalHover,
    colorCriticalActive: tokens_css_1.default.uiKitColorCriticalActive,
    colorCriticalDisabled: tokens_css_1.default.uiKitColorCriticalDisabled,
    colorCriticalLow: tokens_css_1.default.uiKitColorCriticalLow,
    colorCriticalFore: tokens_css_1.default.uiKitColorCriticalFore,
    colorCriticalForeLow: tokens_css_1.default.uiKitColorCriticalForeLow,
    colorCriticalForeDisabled: tokens_css_1.default.uiKitColorCriticalForeDisabled,
    colorCanvas: tokens_css_1.default.uiKitColorCanvas,
    colorTabdock: tokens_css_1.default.uiKitColorTabdock,
    colorPage: tokens_css_1.default.uiKitColorPage,
    colorSurface: tokens_css_1.default.uiKitColorSurface,
    colorTypographyPrimary: tokens_css_1.default.uiKitColorTypographyPrimary,
    colorTypographySecondary: tokens_css_1.default.uiKitColorTypographySecondary,
    colorTypographyTertiary: tokens_css_1.default.uiKitColorTypographyTertiary,
    /**  @deprecated - use `colorTypographyPlaceholder` instead */
    colorTypographyQuaternary: tokens_css_1.default.uiKitColorTypographyQuaternary,
    colorTypographyPlaceholder: tokens_css_1.default.uiKitColorTypographyPlaceholder,
    colorTypographyPositive: tokens_css_1.default.uiKitColorTypographyPositive,
    colorTypographyCritical: tokens_css_1.default.uiKitColorTypographyCritical,
    colorTypographyCriticalHover: tokens_css_1.default.uiKitColorTypographyCriticalHover,
    colorTypographyCriticalActive: tokens_css_1.default.uiKitColorTypographyCriticalActive,
    colorTypographyWarn: tokens_css_1.default.uiKitColorTypographyWarn,
    colorTypographyInfo: tokens_css_1.default.uiKitColorTypographyInfo,
    colorTypographyLink: tokens_css_1.default.uiKitColorTypographyLink,
    colorTypographyLinkHover: tokens_css_1.default.uiKitColorTypographyLinkHover,
    colorTypographyLinkActive: tokens_css_1.default.uiKitColorTypographyLinkActive,
    colorBorder: tokens_css_1.default.uiKitColorBorder,
    colorBorderHover: tokens_css_1.default.uiKitColorBorderHover,
    colorBorderActive: tokens_css_1.default.uiKitColorBorderActive,
    colorBorderDisabled: tokens_css_1.default.uiKitColorBorderDisabled,
    colorBorderCritical: tokens_css_1.default.uiKitColorBorderCritical,
    colorBorderLow: tokens_css_1.default.uiKitColorBorderLow,
    colorBorderStrong: tokens_css_1.default.uiKitColorBorderStrong,
};
