// @canva/app-ui-kit/dist/cjs/ui/base/color_picker/color_picker_flyout.js
"use strict";

const makeObservable = require('../../../base/make_observable/make_observable').makeObservable;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StatelessColorPickerFlyout = exports.ColorPickerFlyout = exports.ColorPickerFlyoutStore = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const mobx_1 = require("mobx");
const mobx_react_1 = require("mobx-react");
const React = require("react");
const box_1 = require('../box/box');
const flyout_1 = require('../surface/flyout/flyout');
const swatch_1 = require('../swatch/swatch');
const color_picker_1 = require("./color_picker");
class ColorPickerFlyoutStore {
  static _makeObservable(instance) {
    makeObservable(instance, {
      open: mobx_1.observable.ref
    });
  }
  constructor() {
    ColorPickerFlyoutStore._makeObservable(this);
    this.open = false;
  }
}
exports.ColorPickerFlyoutStore = ColorPickerFlyoutStore;
const ColorPickerFlyout = ({
  trigger,
  color,
  alpha,
  ariaLabel,
  ariaLabelledBy,
  onChange,
  onChangeStart,
  onChangeComplete,
  onStartEyedropper,
  onOpen,
  onClose
}) => {
  const storeRef = React.useRef(null);
  if (storeRef.current == null) {
    storeRef.current = new ColorPickerFlyoutStore();
  }
  const store = storeRef.current;
  const toggleFlyout = React.useMemo(() => {
    return (0, mobx_1.action)(() => store.open = !store.open);
  }, [store]);
  React.useEffect(() => {
    return (0, mobx_1.reaction)(() => store.open, open => open ? onOpen === null || onOpen === void 0 ? void 0 : onOpen() : onClose === null || onClose === void 0 ? void 0 : onClose());
  }, [store, onOpen, onClose]);
  const triggerId = React.useId();
  return (0, jsx_runtime_1.jsx)(mobx_react_1.Observer, {
    children: () => (0, jsx_runtime_1.jsx)(StatelessColorPickerFlyout, {
      trigger: trigger,
      open: store.open,
      id: triggerId,
      onClick: toggleFlyout,
      ariaLabel: ariaLabel,
      ariaLabelledBy: ariaLabelledBy,
      color: color,
      alpha: alpha,
      onChange: onChange,
      onChangeStart: onChangeStart,
      onChangeComplete: onChangeComplete,
      onStartEyedropper: onStartEyedropper
    })
  });
};
exports.ColorPickerFlyout = ColorPickerFlyout;
function StatelessColorPickerFlyout(props) {
  const {
    trigger = props => (0, jsx_runtime_1.jsx)(swatch_1.AddColorButton, {
      size: "small",
      ...props
    }),
    open,
    id,
    ariaLabel,
    ariaLabelledBy,
    color,
    alpha,
    onClick,
    onChange,
    onChangeStart,
    onChangeComplete,
    onStartEyedropper
  } = props;
  return (0, jsx_runtime_1.jsx)(flyout_1.Flyout, {
    role: "dialog",
    ariaLabelledBy: id,
    placement: "bottom-start",
    open: open,
    onRequestClose: onClick,
    captureFocus: true,
    width: "40u",
    trigger: triggerProps => trigger({
      id,
      ariaLabel,
      ariaLabelledBy,
      onClick,
      ...triggerProps
    }),
    children: ({
      mode
    }) => (0, jsx_runtime_1.jsx)(box_1.Box, {
      padding: mode === 'popover' ? '2u' : '0',
      children: (0, jsx_runtime_1.jsx)(color_picker_1.ColorPicker, {
        sizingMode: mode === 'popover' ? 'fixed' : 'fill',
        color: color,
        alpha: alpha,
        onChange: onChange,
        onChangeStart: onChangeStart,
        onChangeComplete: onChangeComplete,
        onStartEyedropper: onStartEyedropper
      })
    })
  });
}
exports.StatelessColorPickerFlyout = StatelessColorPickerFlyout;
