// @canva/app-ui-kit/dist/cjs/ui/base/color_picker/color_picker_presenter.js
"use strict";

const makeObservable = require('../../../base/make_observable/make_observable').makeObservable;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorPickerPresenter = exports.ColorPickerStores = exports.ColorPickerStore = exports.ColorStoreImpl = void 0;
const tslib_1 = require("tslib");
const preconditions_1 = require('../../../base/preconditions');
const mobx = require("mobx");
const hsv_color_1 = require("./hsv_color/hsv_color");
class ColorStoreImpl {
  static _makeObservable(instance) {
    makeObservable(instance, {
      color: mobx.observable.ref,
      alpha: mobx.observable.ref
    });
  }
  constructor(color, alpha) {
    ColorStoreImpl._makeObservable(this);
    this.color = color;
    this.alpha = alpha;
  }
}
exports.ColorStoreImpl = ColorStoreImpl;
class ColorPickerStore {
  static _makeObservable(instance) {
    makeObservable(instance, {
      lastHsvColor: mobx.observable.ref,
      color: mobx.computed.struct
    });
  }
  constructor(colorStore) {
    ColorPickerStore._makeObservable(this);
    this.colorStore = colorStore;
    this.lastHsvColor = undefined;
  }
  get color() {
    if (this.lastHsvColor == null) {
      const {
        h,
        s,
        v
      } = hsv_color_1.HsvColor.fromHexString(this.colorStore.color);
      return new hsv_color_1.HsvColor(h, s, v, this.colorStore.alpha);
    }
    const {
      h,
      s,
      v,
      alpha
    } = this.lastHsvColor.matchHexString(this.colorStore.color);
    return new hsv_color_1.HsvColor(h, s, v, alpha);
  }
}
exports.ColorPickerStore = ColorPickerStore;
class ColorPickerStores {
  static _makeObservable(instance) {
    makeObservable(instance, {
      lastHsvColor: mobx.observable.ref,
      _color: mobx.observable.ref,
      color: mobx.computed.struct
    });
  }
  constructor(color) {
    ColorPickerStores._makeObservable(this);
    this.lastHsvColor = undefined;
    this._color = color;
  }
  get color() {
    return this.lastHsvColor == null ? hsv_color_1.HsvColor.fromHexString(this._color) : this.lastHsvColor.matchHexString(this._color);
  }
}
exports.ColorPickerStores = ColorPickerStores;
class ColorPickerPresenter {
  static _makeObservable(instance) {
    makeObservable(instance, {
      onInputChange: mobx.action.bound,
      onSatValChange: mobx.action.bound,
      onHueChange: mobx.action.bound,
      onAlphaChange: mobx.action.bound
    });
  }
  constructor(store, onChange) {
    ColorPickerPresenter._makeObservable(this);
    this.store = store;
    this.onChange = onChange;
  }
  onInputChange(hex) {
    this.onChange(hex, 'hex');
  }
  onSatValChange(color) {
    preconditions_1.Preconditions.checkArgument(color.h === this.store.color.h);
    const {
      h,
      s,
      v
    } = color;
    const {
      alpha
    } = this.store.color;
    this.store.lastHsvColor = new hsv_color_1.HsvColor(h, s, v, alpha);
    this.onChange(color.toRgb().toHexString(), 'satVal', alpha);
  }
  onHueChange(hue) {
    const {
      s,
      v,
      alpha
    } = this.store.color;
    const newColor = new hsv_color_1.HsvColor(hue, s, v, alpha);
    this.store.lastHsvColor = newColor;
    if (v > 0 && s > 0) {
      this.onChange(newColor.toRgb().toHexString(), 'hue', alpha);
    }
  }
  onAlphaChange(alpha) {
    const {
      h,
      s,
      v
    } = this.store.color;
    const newColor = new hsv_color_1.HsvColor(h, s, v, alpha);
    this.store.lastHsvColor = newColor;
    this.onChange(newColor.toRgb().toHexString(), 'alpha', alpha);
  }
}
exports.ColorPickerPresenter = ColorPickerPresenter;
