// @canva/app-ui-kit/dist/cjs/ui/base/form/select/base_combo_select.js
"use strict";

const makeObservable = require('../../../../base/make_observable/make_observable').makeObservable;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BaseComboSelect = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const mobx = require("mobx");
const mobx_react_1 = require("mobx-react");
const React = require("react");
const base_input_1 = require('../input/base_input');
const base_select_1 = require("./base_select");
const base_select_presenter_1 = require("./base_select_presenter");
const base_select_util_1 = require("./base_select_util");
const stateless_base_select_1 = require("./stateless_base_select");
class BaseComboSelectStore extends base_select_presenter_1.BaseSelectStore {
  static _makeObservable(instance) {
    makeObservable(instance, {
      value: mobx.observable.ref,
      filteredOptions: mobx.observable.ref
    });
  }
  constructor() {
    super(...arguments);
    BaseComboSelectStore._makeObservable(this);
    this.filteredOptions = [];
  }
}
class BaseComboSelectPresenter extends base_select_presenter_1.BaseSelectPresenter {
  static _makeObservable(instance) {
    makeObservable(instance, {
      filterOptions: mobx.action
    });
  }
  constructor(...args) {
    super(...args);
    BaseComboSelectPresenter._makeObservable(this);
  }
  filterOptions(store, value, options, filterFn) {
    const filteredOptions = filterFn(value, options);
    this.toggle(store, !!filteredOptions.length);
    if (store.open) {
      store.filteredOptions = filteredOptions;
    }
  }
}
const defaultFilterFn = (value, options) => {
  const lowerCaseValue = value.toLocaleLowerCase().replace(/[^\w]/g, '');
  return value ? options.filter(option => {
    var _a;
    const lowerCaseOption = `${(_a = option.value) !== null && _a !== void 0 ? _a : ''} ${option.label}`.toLocaleLowerCase();
    if (lowerCaseOption === lowerCaseValue) {
      return false;
    }
    return lowerCaseOption.replace(/[^\w]/g, '').includes(lowerCaseValue);
  }) : [];
};
let BaseComboSelect = class BaseComboSelect extends React.Component {
  static _makeObservable(instance) {
    makeObservable(instance, {
      onChange: mobx.action
    });
  }
  constructor() {
    super(...arguments);
    BaseComboSelect._makeObservable(this);
    this.presenter = new BaseComboSelectPresenter();
    this.store = new BaseComboSelectStore();
    this.onChange = (value, e) => {
      var _a, _b;
      const {
        filterFn = defaultFilterFn
      } = this.props;
      const options = (0, base_select_util_1.flattenOptions)(this.props.options);
      this.presenter.filterOptions(this.store, value, options, filterFn);
      this.presenter.setValue(this.store, value);
      (_b = (_a = this.props).onChange) === null || _b === void 0 ? void 0 : _b.call(_a, value, e);
    };
    this.Trigger = props => {
      return (0, jsx_runtime_1.jsx)(base_input_1.BaseInput, {
        type: "text",
        value: this.value,
        placeholder: this.props.placeholder,
        disabled: props.disabled,
        error: props.error,
        onChange: this.onChange,
        ariaAutoComplete: "list",
        ariaControls: props.ariaControls,
        ariaActiveDescendant: props.ariaActiveDescendant,
        ariaLabel: props.ariaLabel,
        ariaLabelledBy: props.ariaLabelledBy,
        role: props.role,
        ariaExpanded: props.open
      });
    };
    this.Item = props => {
      return (0, jsx_runtime_1.jsx)(stateless_base_select_1.BaseSelectItem, {
        ...props,
        end: (0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, {})
      });
    };
  }
  get value() {
    var _a;
    return (_a = this.props.value) !== null && _a !== void 0 ? _a : this.store.value;
  }
  render() {
    const {
      value,
      placeholder,
      onChange,
      ...props
    } = this.props;
    return (0, jsx_runtime_1.jsx)(base_select_1.BaseSelect, {
      ...props,
      flyoutMode: "pin",
      Trigger: this.Trigger,
      Item: this.Item,
      presenter: this.presenter,
      store: this.store,
      options: this.store.filteredOptions
    });
  }
};
exports.BaseComboSelect = BaseComboSelect;
exports.BaseComboSelect = BaseComboSelect = tslib_1.__decorate([mobx_react_1.observer], BaseComboSelect);
