// @canva/app-ui-kit/dist/cjs/ui/base/form/select/base_select_presenter.js
"use strict";

const makeObservable = require('../../../../base/make_observable/make_observable').makeObservable;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BaseSelectPresenter = exports.BaseSelectStore = void 0;
const tslib_1 = require("tslib");
const mobx = require("mobx");
class BaseSelectStore {
  static _makeObservable(instance) {
    makeObservable(instance, {
      open: mobx.observable.ref,
      value: mobx.observable.ref,
      query: mobx.observable.ref,
      keyboardNavigation: mobx.observable.ref
    });
  }
  constructor() {
    BaseSelectStore._makeObservable(this);
    this.open = false;
    this.query = '';
  }
}
exports.BaseSelectStore = BaseSelectStore;
class BaseSelectPresenter {
  static _makeObservable(instance) {
    makeObservable(instance, {
      toggle: mobx.action,
      setValue: mobx.action,
      setKeyboardNavigation: mobx.action
    });
  }
  constructor() {
    BaseSelectPresenter._makeObservable(this);
  }
  toggle(store, open) {
    store.open = open != null ? open : !store.open;
  }
  setValue(store, value) {
    store.value = value;
  }
  setKeyboardNavigation(store, value) {
    store.keyboardNavigation = value;
  }
}
exports.BaseSelectPresenter = BaseSelectPresenter;
