// @canva/app-ui-kit/dist/cjs/ui/base/form/select/select.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Select = exports.WindowedSelectMenu = exports.BaseSelectTrigger = exports.BaseSelectMenu = exports.BaseSelectItem = exports.StatelessBaseSelect = exports.BaseSelect = exports.BaseMultiSelect = exports.BaseComboSelect = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const React = require("react");
const base_combo_select_1 = require("./base_combo_select");
const base_multi_select_1 = require("./base_multi_select");
const base_select_1 = require("./base_select");
var base_combo_select_2 = require("./base_combo_select");
Object.defineProperty(exports, "BaseComboSelect", { enumerable: true, get: function () { return base_combo_select_2.BaseComboSelect; } });
var base_multi_select_2 = require("./base_multi_select");
Object.defineProperty(exports, "BaseMultiSelect", { enumerable: true, get: function () { return base_multi_select_2.BaseMultiSelect; } });
var base_select_2 = require("./base_select");
Object.defineProperty(exports, "BaseSelect", { enumerable: true, get: function () { return base_select_2.BaseSelect; } });
var stateless_base_select_1 = require("./stateless_base_select");
Object.defineProperty(exports, "StatelessBaseSelect", { enumerable: true, get: function () { return stateless_base_select_1.StatelessBaseSelect; } });
var stateless_base_select_2 = require("./stateless_base_select");
Object.defineProperty(exports, "BaseSelectItem", { enumerable: true, get: function () { return stateless_base_select_2.BaseSelectItem; } });
Object.defineProperty(exports, "BaseSelectMenu", { enumerable: true, get: function () { return stateless_base_select_2.BaseSelectMenu; } });
Object.defineProperty(exports, "BaseSelectTrigger", { enumerable: true, get: function () { return stateless_base_select_2.BaseSelectTrigger; } });
Object.defineProperty(exports, "WindowedSelectMenu", { enumerable: true, get: function () { return stateless_base_select_2.WindowedSelectMenu; } });
function Select(props) {
    switch (props.type) {
        case 'multi':
            return (0, jsx_runtime_1.jsx)(base_multi_select_1.BaseMultiSelect, { ...props });
        case 'combobox':
            return (0, jsx_runtime_1.jsx)(base_combo_select_1.BaseComboSelect, { ...props });
        default:
            return (0, jsx_runtime_1.jsx)(base_select_1.BaseSelect, { ...props });
    }
}
exports.Select = Select;
