// @canva/app-ui-kit/dist/esm/ui/base/draggable/draggable_view.js
import { makeObservable } from '../../../base/make_observable/make_observable';
import { __decorate } from "tslib";
import * as mobx from 'mobx';
import { observer } from 'mobx-react';
import * as React from 'react';
import { Draggable } from './draggable';
let DraggableView = class DraggableView extends React.Component {
  static _makeObservable(instance) {
    makeObservable(instance, {
      onMouseDown: mobx.action,
      onTouchStart: mobx.action
    });
  }
  constructor() {
    super(...arguments);
    DraggableView._makeObservable(this);
    this.draggable = new Draggable(new Proxy({}, {
      get: (target, p) => this.props[p]
    }));
    this.onMouseDown = e => {
      const {
        onMouseDown
      } = this.props;
      onMouseDown && onMouseDown(e);
      this.draggable.onMouseDown(e);
    };
    this.onTouchStart = e => {
      const {
        onTouchStart
      } = this.props;
      onTouchStart && onTouchStart(e);
      this.draggable.onTouchStart(e);
    };
  }
  componentWillUnmount() {
    if (!this.props.continueAfterUnmount) {
      this.draggable.stopInProgressDrag();
    }
  }
  render() {
    return this.props.children({
      onMouseDown: this.onMouseDown,
      onTouchStart: this.onTouchStart
    });
  }
};
DraggableView = __decorate([observer], DraggableView);
export { DraggableView };
