// @canva/app-ui-kit/dist/esm/ui/base/placeholder/placeholder.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { useEnableAnimations } from '../provider/provider';
import styles, { getStyle } from './placeholder.css';
export const Placeholder = ({ shape, index, disableAnimations }) => {
    const className = classNames(styles.base, {
        [styles.animated]: useEnableAnimations() && disableAnimations !== true,
    }, getStyle(shape));
    const style = {};
    if (index != null) {
        style.animationDelay = `calc(${index} * ${styles.flashDelayDuration})`;
    }
    return _jsx("div", { className: className, style: style });
};
export const CirclePlaceholder = props => (_jsx(Placeholder, { shape: "circle", ...props }));
export const SquarePlaceholder = props => (_jsx(Placeholder, { shape: "square", ...props }));
export const RectanglePlaceholder = props => (_jsx(Placeholder, { shape: "rectangle", ...props }));
export const SharpRectanglePlaceholder = props => (_jsx(Placeholder, { shape: "sharpRectangle", ...props }));
export const TextPlaceholder = props => (_jsx(Placeholder, { shape: "text", ...props }));
export const TitlePlaceholder = props => (_jsx(Placeholder, { shape: "title", ...props }));
