// @canva/app-ui-kit/dist/cjs/ui/base/card/internal/card_context.js
"use strict";

const makeObservable = require('../../../../base/make_observable/make_observable').makeObservable;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withCardContextProvider = exports.useCardEventHandlers = exports.useCardMouseDown = exports.useCardFocus = exports.useCardHover = exports.useCardConfigurationSetter = exports.useCardConfiguration = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const mobx = require("mobx");
const React = require("react");
const useCardConfiguration = () => useCardConfigurationStore().configuration;
exports.useCardConfiguration = useCardConfiguration;
function useCardConfigurationSetter(configuration) {
  const {
    setConfiguration
  } = useCardConfigurationStore();
  const willMount = React.useRef(true);
  if (willMount.current) {
    setConfiguration(configuration);
  }
  willMount.current = false;
  React.useEffect(() => {
    setConfiguration(configuration);
  }, [setConfiguration, configuration]);
}
exports.useCardConfigurationSetter = useCardConfigurationSetter;
function useCardHover(args) {
  var _a;
  const track = (_a = args === null || args === void 0 ? void 0 : args.track) !== null && _a !== void 0 ? _a : true;
  const interactionStore = useCardInteractionStore();
  React.useEffect(() => {
    if (track) {
      mobx.runInAction(() => interactionStore.trackHover = true);
    }
  }, [interactionStore, track]);
  return interactionStore.isHovered;
}
exports.useCardHover = useCardHover;
function useCardFocus() {
  const interactionStore = useCardInteractionStore();
  React.useEffect(() => {
    mobx.runInAction(() => interactionStore.trackFocus = true);
  }, [interactionStore]);
  return interactionStore.isFocused;
}
exports.useCardFocus = useCardFocus;
function useCardMouseDown() {
  const interactionStore = useCardInteractionStore();
  React.useEffect(() => {
    mobx.runInAction(() => interactionStore.trackMouseDown = true);
  }, [interactionStore]);
  return interactionStore.isMouseDown;
}
exports.useCardMouseDown = useCardMouseDown;
const useCardEventHandlers = () => useCardInteractionStore().eventHandlers;
exports.useCardEventHandlers = useCardEventHandlers;
function withCardContextProvider(Component) {
  var _a;
  const WithProvider = props => {
    const [context] = React.useState(createContext);
    return (0, jsx_runtime_1.jsx)(CardContext.Provider, {
      value: context,
      children: (0, jsx_runtime_1.jsx)(Component, {
        ...props
      })
    });
  };
  WithProvider.displayName = `withCardContext(${(_a = Component.displayName) !== null && _a !== void 0 ? _a : Component.name})`;
  return WithProvider;
}
exports.withCardContextProvider = withCardContextProvider;
class CardInteractionStore {
  static _makeObservable(instance) {
    makeObservable(instance, {
      isHovered: mobx.observable.ref,
      isFocused: mobx.observable.ref,
      isMouseDown: mobx.observable.ref,
      trackHover: mobx.observable.ref,
      trackFocus: mobx.observable.ref,
      trackMouseDown: mobx.observable.ref,
      eventHandlers: mobx.computed,
      setHovered: mobx.action.bound,
      setFocused: mobx.action.bound,
      setMouseDown: mobx.action.bound
    });
  }
  constructor() {
    CardInteractionStore._makeObservable(this);
    this.isHovered = false;
    this.isFocused = false;
    this.isMouseDown = false;
    this.trackHover = false;
    this.trackFocus = false;
    this.trackMouseDown = false;
  }
  get eventHandlers() {
    return {
      onMouseEnter: this.trackHover ? () => this.setHovered(true) : undefined,
      onMouseLeave: this.trackHover ? () => this.setHovered(false) : undefined,
      onFocus: this.trackFocus ? () => this.setFocused(true) : undefined,
      onBlur: this.trackFocus ? () => this.setFocused(false) : undefined,
      onMouseDown: this.trackMouseDown ? () => this.setMouseDown(true) : undefined,
      onMouseUp: this.trackMouseDown ? () => this.setMouseDown(false) : undefined
    };
  }
  setHovered(isHovered) {
    this.isHovered = isHovered;
  }
  setFocused(isFocused) {
    this.isFocused = isFocused;
  }
  setMouseDown(isMouseDown) {
    this.isMouseDown = isMouseDown;
  }
}
class CardConfigurationStore {
  static _makeObservable(instance) {
    makeObservable(instance, {
      configuration: mobx.observable.struct,
      setConfiguration: mobx.action.bound
    });
  }
  constructor() {
    CardConfigurationStore._makeObservable(this);
    this.configuration = {
      size: 'medium',
      sharp: false,
      border: true,
      disabled: false,
      badgeTransition: 'fade',
      hasHoverEffect: true
    };
  }
  setConfiguration(configuration) {
    this.configuration = configuration;
  }
}
function createContext() {
  return {
    interactionStore: new CardInteractionStore(),
    configurationStore: new CardConfigurationStore()
  };
}
const CardContext = React.createContext(undefined);
const useCardContext = () => {
  var _a;
  return (_a = React.useContext(CardContext)) !== null && _a !== void 0 ? _a : createContext();
};
const useCardConfigurationStore = () => useCardContext().configurationStore;
const useCardInteractionStore = () => useCardContext().interactionStore;
