// @canva/app-ui-kit/dist/cjs/ui/base/placeholder/placeholder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TitlePlaceholder = exports.TextPlaceholder = exports.SharpRectanglePlaceholder = exports.RectanglePlaceholder = exports.SquarePlaceholder = exports.CirclePlaceholder = exports.Placeholder = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const classnames_1 = require("classnames");
const React = require("react");
const provider_1 = require('../provider/provider');
const placeholder_css_1 = require("./placeholder.css");
const Placeholder = ({ shape, index, disableAnimations }) => {
    const className = (0, classnames_1.default)(placeholder_css_1.default.base, {
        [placeholder_css_1.default.animated]: (0, provider_1.useEnableAnimations)() && disableAnimations !== true,
    }, (0, placeholder_css_1.getStyle)(shape));
    const style = {};
    if (index != null) {
        style.animationDelay = `calc(${index} * ${placeholder_css_1.default.flashDelayDuration})`;
    }
    return (0, jsx_runtime_1.jsx)("div", { className: className, style: style });
};
exports.Placeholder = Placeholder;
const CirclePlaceholder = props => ((0, jsx_runtime_1.jsx)(exports.Placeholder, { shape: "circle", ...props }));
exports.CirclePlaceholder = CirclePlaceholder;
const SquarePlaceholder = props => ((0, jsx_runtime_1.jsx)(exports.Placeholder, { shape: "square", ...props }));
exports.SquarePlaceholder = SquarePlaceholder;
const RectanglePlaceholder = props => ((0, jsx_runtime_1.jsx)(exports.Placeholder, { shape: "rectangle", ...props }));
exports.RectanglePlaceholder = RectanglePlaceholder;
const SharpRectanglePlaceholder = props => ((0, jsx_runtime_1.jsx)(exports.Placeholder, { shape: "sharpRectangle", ...props }));
exports.SharpRectanglePlaceholder = SharpRectanglePlaceholder;
const TextPlaceholder = props => ((0, jsx_runtime_1.jsx)(exports.Placeholder, { shape: "text", ...props }));
exports.TextPlaceholder = TextPlaceholder;
const TitlePlaceholder = props => ((0, jsx_runtime_1.jsx)(exports.Placeholder, { shape: "title", ...props }));
exports.TitlePlaceholder = TitlePlaceholder;
