// @canva/app-ui-kit/dist/esm/ui/base/form/select/base_select_presenter.js
import { makeObservable } from '../../../../base/make_observable/make_observable';
import { __decorate } from "tslib";
import * as mobx from 'mobx';
export class BaseSelectStore {
  static _makeObservable(instance) {
    makeObservable(instance, {
      open: mobx.observable.ref,
      value: mobx.observable.ref,
      query: mobx.observable.ref,
      keyboardNavigation: mobx.observable.ref
    });
  }
  constructor() {
    BaseSelectStore._makeObservable(this);
    this.open = false;
    this.query = '';
  }
}
export class BaseSelectPresenter {
  static _makeObservable(instance) {
    makeObservable(instance, {
      toggle: mobx.action,
      setValue: mobx.action,
      setKeyboardNavigation: mobx.action
    });
  }
  constructor() {
    BaseSelectPresenter._makeObservable(this);
  }
  toggle(store, open) {
    store.open = open != null ? open : !store.open;
  }
  setValue(store, value) {
    store.value = value;
  }
  setKeyboardNavigation(store, value) {
    store.keyboardNavigation = value;
  }
}
