// @canva/app-ui-kit/dist/cjs/ui/base/draggable/touch_tracker.js
"use strict";

const makeObservable = require('../../../base/make_observable/make_observable').makeObservable;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TouchTracker = void 0;
const tslib_1 = require("tslib");
const mobx = require("mobx");
const client_position_1 = require("./client_position");
const drag_handler_1 = require("./drag_handler");
class TouchTracker {
  static _makeObservable(instance) {
    makeObservable(instance, {
      onTouchStart: mobx.action.bound,
      onTouchMove: mobx.action.bound,
      onTouchEnd: mobx.action.bound
    });
  }
  constructor(drag, touch, holdToDrag, onStopped) {
    TouchTracker._makeObservable(this);
    this.drag = drag;
    this.holdToDrag = holdToDrag;
    this.onStopped = onStopped;
    this.type = 'touch';
    this.identifier = touch.identifier;
    this.target = touch.target;
    this.target.addEventListener('touchmove', this.onTouchMove);
    this.target.addEventListener('touchend', this.onTouchEnd);
    this.target.addEventListener('touchcancel', this.onTouchEnd);
    document.addEventListener('touchstart', this.onTouchStart);
  }
  stop() {
    try {
      this.drag.stop();
    } finally {
      this.target.removeEventListener('touchmove', this.onTouchMove);
      this.target.removeEventListener('touchend', this.onTouchEnd);
      this.target.removeEventListener('touchcancel', this.onTouchEnd);
      document.removeEventListener('touchstart', this.onTouchStart);
      this.onStopped();
    }
  }
  getChangedTouch(evt) {
    for (const touch of evt.changedTouches) {
      if (touch.identifier === this.identifier) {
        return touch;
      }
    }
  }
  onTouchStart(e) {
    if (e.touches.length !== 1 || e.changedTouches[0].identifier !== this.identifier) {
      this.stop();
    }
  }
  onTouchMove(e) {
    const touch = this.getChangedTouch(e);
    if (touch) {
      if (!this.holdToDrag || this.drag.started) {
        e.preventDefault();
      }
      this.drag.handleMove(client_position_1.ClientPosition.fromEvent(touch), drag_handler_1.DragHandler.modifiersFromEvent(e));
    }
  }
  onTouchEnd(e) {
    const touch = this.getChangedTouch(e);
    if (touch) {
      try {
        this.drag.updateAndMayStart(client_position_1.ClientPosition.fromEvent(touch), drag_handler_1.DragHandler.modifiersFromEvent(e));
      } finally {
        this.stop();
      }
    }
  }
}
exports.TouchTracker = TouchTracker;
