// @canva/app-ui-kit/dist/esm/ui/base/card/internal/card_thumbnail.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { delay as delayFn } from '../../../../base/delay';
import classNames from 'classnames';
import { action, autorun } from 'mobx';
import { observer, useLocalObservable } from 'mobx-react';
import * as React from 'react';
import { usePrefersReducedMotion } from '../../animation/supports_animation';
import { useAutoplayVideos } from '../../provider/provider';
import { IconThumbnail } from '../../thumbnail/icon_thumbnail/icon_thumbnail';
import { Image as BaseImage } from '../../thumbnail/image/image';
import { ImagePlaceholder, ImageThumbnail, } from '../../thumbnail/image_thumbnail/image_thumbnail';
import { Video } from '../../video/video';
import styles from '../card.css';
import { useCardConfiguration, useCardFocus, useCardHover } from './card_context';
const TRANSPARENT_IMAGE = 'data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=';
export const CardPlaceholder = ImagePlaceholder;
export const CardIconThumbnail = (props) => (_jsx(IconThumbnail, { Icon: props.Icon, size: props.size, tone: "secondary" }));
export const CardImageThumbnail = React.memo(function CardImageThumbnail(props) {
    const { ImageComponent = BaseImage } = props;
    const { sharp } = useCardConfiguration();
    return (_jsx(ImageThumbnail, { ...props, src: props.url, border: false, ImageComponent: ({ ...imageProps }) => {
            var _a;
            return (_jsx(ImageComponent, { ...imageProps, className: classNames(props.className, styles.image, {
                    [styles.roundBorders]: !((_a = props.sharp) !== null && _a !== void 0 ? _a : sharp),
                }), crossOrigin: props.crossOrigin, src: props.url, alt: props.alt || '', elementTiming: props.elementTiming }));
        } }));
});
CardImageThumbnail.displayName = 'CardImageThumbnail';
export const CardVideoThumbnail = observer(React.forwardRef(function CardVideoThumbnail(props, ref) {
    var _a;
    const { className, loadVideo = simpleLoadVideo, VideoComponent = Video, ImageComponent, videoUrl, videoLoadingState, imageUrl, imageLoadingState, imageFallback, imagePlaceholder, imageAlt, } = props;
    const config = useCardConfiguration();
    const isFocused = useCardFocus();
    const autoplayVideos = useAutoplayVideos() || 'ADAPTIVE';
    const prefersReducedMotion = usePrefersReducedMotion();
    const enableAutoplay = (!prefersReducedMotion && autoplayVideos === 'ADAPTIVE') || autoplayVideos === 'AUTOPLAY';
    const sharp = (_a = props.sharp) !== null && _a !== void 0 ? _a : config.sharp;
    const whenToPlay = imageUrl == null
        ? 'always'
        : props.whenToPlay === 'always' && enableAutoplay
            ? props.whenToPlay
            : 'on-hover';
    const isHovered = useCardHover({ track: whenToPlay === 'on-hover' });
    const store = useLocalObservable(() => ({
        videoLoadingState: 'loading',
        imageLoadingState: imageUrl == null ? 'loaded' : imageLoadingState == null ? 'loading' : imageLoadingState,
        playState: 'pause',
        setPlayState: action((value) => {
            store.playState = value;
        }),
        setImageLoadingState: action((imageLoadingState) => {
            store.imageLoadingState = imageLoadingState;
        }),
        setVideoLoadingState: action((videoLoadingState) => {
            store.videoLoadingState = videoLoadingState;
        }),
    }));
    const { setPlayState } = store;
    React.useEffect(() => autorun(() => {
        if (whenToPlay === 'always' && store.playState !== 'play') {
            setPlayState('play');
        }
    }), [whenToPlay, store.playState, setPlayState]);
    React.useEffect(() => autorun(() => {
        if (whenToPlay === 'on-hover') {
            if ((isHovered || isFocused) && store.playState !== 'play') {
                setPlayState('play');
            }
            else if (!isHovered && !isFocused && store.playState === 'play') {
                setPlayState('pause');
            }
        }
    }), [whenToPlay, isHovered, isFocused, store.playState, setPlayState]);
    const renderVideoThumbnail = React.useCallback((videoUrl_) => {
        const renderVideo = (loadingState) => loadingState === 'loaded' && (_jsx(VideoComponent, { autoPlay: true, className: classNames(styles.video, className, {
                [styles.roundBorders]: !(sharp !== null && sharp !== void 0 ? sharp : config.sharp),
            }), controls: false, draggable: false, loop: true, muted: true, playsInline: true, ref: ref, src: videoUrl_, 
            poster: isIosCanvaEditor() ? TRANSPARENT_IMAGE : undefined }));
        if (videoLoadingState == null && store.videoLoadingState === 'loading') {
            const delay = whenToPlay === 'on-hover' ? 150 : 0;
            return (_jsx(LoadVideo, { url: videoUrl_, delay: delay, loadVideo: loadVideo, callback: store.setVideoLoadingState, children: loadingState => renderVideo(loadingState) }));
        }
        return renderVideo(videoLoadingState || store.videoLoadingState);
    }, [
        ref,
        videoLoadingState,
        store.videoLoadingState,
        whenToPlay,
        sharp,
        config.sharp,
        className,
        VideoComponent,
        loadVideo,
        store.setVideoLoadingState,
    ]);
    const renderImageThumbnail = React.useCallback((imageUrl_) => (_jsx(CardImageThumbnail, { url: imageUrl_, loadingState: imageLoadingState, fallback: imageFallback, placeholder: imagePlaceholder, sharp: sharp, className: className, ImageComponent: ImageComponent, alt: imageAlt, onImageLoad: store.setImageLoadingState })), [
        imageLoadingState,
        imageFallback,
        imagePlaceholder,
        store.setImageLoadingState,
        className,
        sharp,
        ImageComponent,
        imageAlt,
    ]);
    return (_jsxs(_Fragment, { children: [imageUrl != null && renderImageThumbnail(imageUrl), store.playState === 'play'
                && ((imageLoadingState && imageLoadingState !== 'loading')
                    || store.imageLoadingState !== 'loading')
                && renderVideoThumbnail(videoUrl)] }));
}));
function simpleLoadVideo(url) {
    return new Promise((resolve, reject) => {
        const onLoad = () => {
            video.removeEventListener('loadeddata', onLoad);
            resolve(video);
        };
        const onError = () => {
            video.removeEventListener('error', onError);
            reject();
        };
        const video = document.createElement('video');
        video.src = url;
        video.addEventListener('error', onError);
        video.addEventListener('loadeddata', onLoad);
        video.load();
    });
}
const LoadVideo = observer(function LoadVideo({ url, delay, callback, loadVideo = simpleLoadVideo, children, }) {
    const store = useLocalObservable(() => ({
        loadingState: 'loading',
        setLoadingState: action((loadingState) => {
            store.loadingState = loadingState;
        }),
    }));
    React.useEffect(() => {
        (async () => {
            await delayFn(delay);
            callback && callback('loading');
            await loadVideo(url)
                .then(() => {
                store.setLoadingState('loaded');
                callback && callback('loaded');
            })
                .catch(() => {
                store.setLoadingState('error');
                callback && callback('error');
            });
        })();
    }, [url, store, callback, loadVideo, delay]);
    return _jsx(_Fragment, { children: children(store.loadingState) });
});
const isIosCanvaEditor = () => {
    return (navigator != null
        && navigator.userAgent.indexOf('canvaeditor') >= 0 && !!navigator.userAgent.match(/iphone|ipad/gi));
};
