// @canva/app-ui-kit/dist/esm/ui/base/a11y/access_mode_state/access_mode_state.js
import { makeObservable } from '../../../../base/make_observable/make_observable';
import { __decorate } from "tslib";
import { action, computed, observable, reaction } from 'mobx';
import * as React from 'react';
const initialAccessMode = 'mouse';
class AccessModeStore {
  static _makeObservable(instance) {
    makeObservable(instance, {
      _accessMode: observable.ref,
      isMouseMode: computed,
      isKeyboardMode: computed,
      accessMode: computed,
      setAccessMode: action
    });
  }
  constructor() {
    AccessModeStore._makeObservable(this);
    this._accessMode = initialAccessMode;
  }
  get isMouseMode() {
    return this._accessMode === 'mouse';
  }
  get isKeyboardMode() {
    return this._accessMode === 'keyboard';
  }
  get accessMode() {
    return this._accessMode;
  }
  setAccessMode(mode) {
    this._accessMode = mode;
    return this._accessMode;
  }
  resetAccessMode() {
    this.setAccessMode(initialAccessMode);
  }
}
export const accessModeState = new AccessModeStore();
function registerDocumentListeners(document) {
  document.addEventListener('keydown', e => {
    if (e.key === 'Tab') {
      accessModeState.setAccessMode('keyboard');
    }
  }, true);
  document.addEventListener('mousedown', e => accessModeState.setAccessMode('mouse'), true);
  document.addEventListener('touchstart', e => accessModeState.setAccessMode('mouse'), true);
}
if (typeof document !== 'undefined') {
  registerDocumentListeners(document);
}
function useReaction(expression, initialValue) {
  const [state, setState] = React.useState(initialValue);
  React.useEffect(() => {
    return reaction(expression, value => setState(value), {
      fireImmediately: true
    });
  }, []);
  return state;
}
export function useIsMouseAccessMode() {
  return useReaction(() => accessModeState.isMouseMode, true);
}
